{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import piezo\n",
    "\n",
    "%load_ext autoreload\n",
    "\n",
    "%autoreload 2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "cat=piezo.ResistanceCatalogue(\"config/NC_00962.3_NEJM2018_GM3_RUS_v1.0.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-81-dff39c265e99>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcat\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'rpoB_S450L'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/packages/piezo/piezo/catalogue.py\u001b[0m in \u001b[0;36mpredict\u001b[0;34m(self, mutation, verbose)\u001b[0m\n\u001b[1;32m     28\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 30\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcatalogue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmutation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mverbose\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     31\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mparse_json\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/packages/piezo/piezo/catalogue.py\u001b[0m in \u001b[0;36mpredict\u001b[0;34m(catalogue, mutation, verbose)\u001b[0m\n\u001b[1;32m    153\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcatalogue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgrammar\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"GM1\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    154\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 155\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpredict_GM1\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcatalogue\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmutation\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mverbose\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    156\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    157\u001b[0m         \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/packages/piezo/piezo/catalogue.py\u001b[0m in \u001b[0;36mpredict_GM1\u001b[0;34m(catalogue, gene_mutation, verbose)\u001b[0m\n\u001b[1;32m    215\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mmutation_type\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"SNP\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    216\u001b[0m                 process_snp_variants(mutation_affects, predictions, catalogue, before, after, mutation, rules_subset,\n\u001b[0;32m--> 217\u001b[0;31m                                      mutation_type_subset_vector, position_subset_vector, verbose)\n\u001b[0m\u001b[1;32m    218\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mmutation_type\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m\"INDEL\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    219\u001b[0m                 process_indel_variants(variant_affects,\n",
      "\u001b[0;32m~/packages/piezo/piezo/catalogue.py\u001b[0m in \u001b[0;36mprocess_snp_variants\u001b[0;34m(mutation_affects, predictions, catalogue, before, after, mutation, rules, rules_mutation_type_vector, rules_position_vector, verbose)\u001b[0m\n\u001b[1;32m    268\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    269\u001b[0m     \u001b[0;31m# PRIORITY=5: any het mutation at this specific position in the CDS or PROM  (e.g. rpoB_S450Z or rpoB_c-15z)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 270\u001b[0;31m     \u001b[0mrow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrules\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrules_mutation_type_vector\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mrules_position_vector\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mbefore\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0mafter\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmutation\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"Z\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"z\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m&\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mrules\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mMUTATION\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'Z'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'z'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    271\u001b[0m     \u001b[0mrow_prediction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpredictions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"het SNP at specified position in the CDS\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    272\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Library/Python/3.5/lib/python/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__nonzero__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1553\u001b[0m             \u001b[0;34m\"The truth value of a {0} is ambiguous. \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1554\u001b[0m             \"Use a.empty, a.bool(), a.item(), a.any() or a.all().\".format(\n\u001b[0;32m-> 1555\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1556\u001b[0m             )\n\u001b[1;32m   1557\u001b[0m         )\n",
      "\u001b[0;31mValueError\u001b[0m: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all()."
     ]
    }
   ],
   "source": [
    "cat.predict('rpoB_S450L')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG</th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>PREDICTION</th>\n",
       "      <th>EVIDENCE</th>\n",
       "      <th>SOURCE</th>\n",
       "      <th>OTHER</th>\n",
       "      <th>GENE</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>MUTATION_AFFECTS</th>\n",
       "      <th>MUTATION_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>EMB</td>\n",
       "      <td>2723_indel</td>\n",
       "      <td>S</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'PUBLICATION': 'LID2015'}</td>\n",
       "      <td>{'PHYLO_SNP': False}</td>\n",
       "      <td>embA</td>\n",
       "      <td>2723</td>\n",
       "      <td>CDS</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>EMB</td>\n",
       "      <td>A1015T</td>\n",
       "      <td>S</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'PUBLICATION': 'LID2015'}</td>\n",
       "      <td>{'PHYLO_SNP': False}</td>\n",
       "      <td>embA</td>\n",
       "      <td>1015</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>EMB</td>\n",
       "      <td>A1016S</td>\n",
       "      <td>S</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'PUBLICATION': 'LID2015'}</td>\n",
       "      <td>{'PHYLO_SNP': False}</td>\n",
       "      <td>embA</td>\n",
       "      <td>1016</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>EMB</td>\n",
       "      <td>A109T</td>\n",
       "      <td>S</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'PUBLICATION': 'LID2015'}</td>\n",
       "      <td>{'PHYLO_SNP': False}</td>\n",
       "      <td>embA</td>\n",
       "      <td>109</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>EMB</td>\n",
       "      <td>A201T</td>\n",
       "      <td>S</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'PUBLICATION': 'LID2015'}</td>\n",
       "      <td>{'PHYLO_SNP': False}</td>\n",
       "      <td>embA</td>\n",
       "      <td>201</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1146</td>\n",
       "      <td>PZA</td>\n",
       "      <td>*?</td>\n",
       "      <td>U</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'PUBLICATION': 'LID2015'}</td>\n",
       "      <td>{}</td>\n",
       "      <td>panD</td>\n",
       "      <td>*</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1147</td>\n",
       "      <td>PZA</td>\n",
       "      <td>*=</td>\n",
       "      <td>S</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'PUBLICATION': 'LID2015'}</td>\n",
       "      <td>{}</td>\n",
       "      <td>panD</td>\n",
       "      <td>*</td>\n",
       "      <td>CDS</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1148</td>\n",
       "      <td>PZA</td>\n",
       "      <td>-*?</td>\n",
       "      <td>U</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'PUBLICATION': 'LID2015'}</td>\n",
       "      <td>{}</td>\n",
       "      <td>panD</td>\n",
       "      <td>-*</td>\n",
       "      <td>PROM</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1149</td>\n",
       "      <td>PZA</td>\n",
       "      <td>*_indel</td>\n",
       "      <td>U</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'PUBLICATION': 'LID2015'}</td>\n",
       "      <td>{}</td>\n",
       "      <td>panD</td>\n",
       "      <td>*</td>\n",
       "      <td>CDS</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1150</td>\n",
       "      <td>PZA</td>\n",
       "      <td>-*_indel</td>\n",
       "      <td>U</td>\n",
       "      <td>{}</td>\n",
       "      <td>{'PUBLICATION': 'LID2015'}</td>\n",
       "      <td>{}</td>\n",
       "      <td>panD</td>\n",
       "      <td>-*</td>\n",
       "      <td>PROM</td>\n",
       "      <td>INDEL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1151 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     DRUG    MUTATION PREDICTION EVIDENCE                      SOURCE  \\\n",
       "0     EMB  2723_indel          S       {}  {'PUBLICATION': 'LID2015'}   \n",
       "1     EMB      A1015T          S       {}  {'PUBLICATION': 'LID2015'}   \n",
       "2     EMB      A1016S          S       {}  {'PUBLICATION': 'LID2015'}   \n",
       "3     EMB       A109T          S       {}  {'PUBLICATION': 'LID2015'}   \n",
       "4     EMB       A201T          S       {}  {'PUBLICATION': 'LID2015'}   \n",
       "...   ...         ...        ...      ...                         ...   \n",
       "1146  PZA          *?          U       {}  {'PUBLICATION': 'LID2015'}   \n",
       "1147  PZA          *=          S       {}  {'PUBLICATION': 'LID2015'}   \n",
       "1148  PZA         -*?          U       {}  {'PUBLICATION': 'LID2015'}   \n",
       "1149  PZA     *_indel          U       {}  {'PUBLICATION': 'LID2015'}   \n",
       "1150  PZA    -*_indel          U       {}  {'PUBLICATION': 'LID2015'}   \n",
       "\n",
       "                     OTHER  GENE POSITION MUTATION_AFFECTS MUTATION_TYPE  \n",
       "0     {'PHYLO_SNP': False}  embA     2723              CDS         INDEL  \n",
       "1     {'PHYLO_SNP': False}  embA     1015              CDS           SNP  \n",
       "2     {'PHYLO_SNP': False}  embA     1016              CDS           SNP  \n",
       "3     {'PHYLO_SNP': False}  embA      109              CDS           SNP  \n",
       "4     {'PHYLO_SNP': False}  embA      201              CDS           SNP  \n",
       "...                    ...   ...      ...              ...           ...  \n",
       "1146                    {}  panD        *              CDS           SNP  \n",
       "1147                    {}  panD        *              CDS           SNP  \n",
       "1148                    {}  panD       -*             PROM           SNP  \n",
       "1149                    {}  panD        *              CDS         INDEL  \n",
       "1150                    {}  panD       -*             PROM         INDEL  \n",
       "\n",
       "[1151 rows x 10 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat.catalogue.rules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "predict() missing 1 required positional argument: 'mutation'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-45-48d2c01214f0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpiezo\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mResistanceCatalogue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: predict() missing 1 required positional argument: 'mutation'"
     ]
    }
   ],
   "source": [
    "piezo.ResistanceCatalogue.predict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
