
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../catalogue/">
      
      
        <link rel="next" href="../../nomenclature/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.3">
    
    
      
        <title>grammar_GARC1 - Piezo Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.d7758b05.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#grammar_GARC1" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Piezo Documentation" class="md-header__button md-logo" aria-label="Piezo Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Piezo Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              grammar_GARC1
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/oxfordmmm/piezo" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    piezo
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../readme/" class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../" class="md-tabs__link">
          
  
  Code Reference

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../nomenclature/" class="md-tabs__link">
        
  
    
  
  Nomenclature

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Piezo Documentation" class="md-nav__button md-logo" aria-label="Piezo Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Piezo Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/oxfordmmm/piezo" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    piezo
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../readme/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Code Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    piezo
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../catalogue/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    catalogue
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    grammar_GARC1
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    grammar_GARC1
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1" class="md-nav__link">
    <span class="md-ellipsis">
      grammar_GARC1
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.infer_mutation_affects" class="md-nav__link">
    <span class="md-ellipsis">
      infer_mutation_affects
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.large_del" class="md-nav__link">
    <span class="md-ellipsis">
      large_del
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.match_multi" class="md-nav__link">
    <span class="md-ellipsis">
      match_multi
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.merge_predictions" class="md-nav__link">
    <span class="md-ellipsis">
      merge_predictions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.parse_mutation" class="md-nav__link">
    <span class="md-ellipsis">
      parse_mutation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.predict" class="md-nav__link">
    <span class="md-ellipsis">
      predict
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.predict_GARC1" class="md-nav__link">
    <span class="md-ellipsis">
      predict_GARC1
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.predict_multi" class="md-nav__link">
    <span class="md-ellipsis">
      predict_multi
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.process_catalogue_GARC1" class="md-nav__link">
    <span class="md-ellipsis">
      process_catalogue_GARC1
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.process_indel_variants" class="md-nav__link">
    <span class="md-ellipsis">
      process_indel_variants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.process_snp_variants" class="md-nav__link">
    <span class="md-ellipsis">
      process_snp_variants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.row_prediction" class="md-nav__link">
    <span class="md-ellipsis">
      row_prediction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.sanity_check_snp" class="md-nav__link">
    <span class="md-ellipsis">
      sanity_check_snp
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.split_mutation" class="md-nav__link">
    <span class="md-ellipsis">
      split_mutation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.validate_multi" class="md-nav__link">
    <span class="md-ellipsis">
      validate_multi
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../nomenclature/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Nomenclature
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1" class="md-nav__link">
    <span class="md-ellipsis">
      grammar_GARC1
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.infer_mutation_affects" class="md-nav__link">
    <span class="md-ellipsis">
      infer_mutation_affects
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.large_del" class="md-nav__link">
    <span class="md-ellipsis">
      large_del
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.match_multi" class="md-nav__link">
    <span class="md-ellipsis">
      match_multi
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.merge_predictions" class="md-nav__link">
    <span class="md-ellipsis">
      merge_predictions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.parse_mutation" class="md-nav__link">
    <span class="md-ellipsis">
      parse_mutation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.predict" class="md-nav__link">
    <span class="md-ellipsis">
      predict
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.predict_GARC1" class="md-nav__link">
    <span class="md-ellipsis">
      predict_GARC1
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.predict_multi" class="md-nav__link">
    <span class="md-ellipsis">
      predict_multi
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.process_catalogue_GARC1" class="md-nav__link">
    <span class="md-ellipsis">
      process_catalogue_GARC1
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.process_indel_variants" class="md-nav__link">
    <span class="md-ellipsis">
      process_indel_variants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.process_snp_variants" class="md-nav__link">
    <span class="md-ellipsis">
      process_snp_variants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.row_prediction" class="md-nav__link">
    <span class="md-ellipsis">
      row_prediction
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.sanity_check_snp" class="md-nav__link">
    <span class="md-ellipsis">
      sanity_check_snp
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.split_mutation" class="md-nav__link">
    <span class="md-ellipsis">
      split_mutation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#grammar_GARC1.validate_multi" class="md-nav__link">
    <span class="md-ellipsis">
      validate_multi
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>grammar_GARC1</h1>

<div class="doc doc-object doc-module">



<a id="grammar_GARC1"></a>
    <div class="doc doc-contents first">

        <p>All logic required to parse and generate predictions from a catalogue in GARC1</p>








  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.infer_mutation_affects" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">infer_mutation_affects</span><span class="p">(</span><span class="n">position</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Find out which part of the gene this position affects</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>position</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene position</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>str</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Either "PROM" or "CDS"</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1351</span>
<span class="normal">1352</span>
<span class="normal">1353</span>
<span class="normal">1354</span>
<span class="normal">1355</span>
<span class="normal">1356</span>
<span class="normal">1357</span>
<span class="normal">1358</span>
<span class="normal">1359</span>
<span class="normal">1360</span>
<span class="normal">1361</span>
<span class="normal">1362</span>
<span class="normal">1363</span>
<span class="normal">1364</span>
<span class="normal">1365</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">infer_mutation_affects</span><span class="p">(</span><span class="n">position</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Find out which part of the gene this position affects</span>

<span class="sd">    Args:</span>
<span class="sd">        position (int): Gene position</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Either &quot;PROM&quot; or &quot;CDS&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">position</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">mutation_affects</span> <span class="o">=</span> <span class="s2">&quot;PROM&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">mutation_affects</span> <span class="o">=</span> <span class="s2">&quot;CDS&quot;</span>
    <span class="k">return</span> <span class="n">mutation_affects</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.large_del" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">large_del</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">rules</span><span class="p">,</span> <span class="n">size</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Row prediction, but specifically for large deletions</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>predictions</code>
            </td>
            <td>
                  <code>{int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str}): Predictions dictionary</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rules</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Rules from the catalogue</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>size</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Percentage of the gene deleted (as a decimal)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>minor</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minors (if existing)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span>
<span class="normal">880</span>
<span class="normal">881</span>
<span class="normal">882</span>
<span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">large_del</span><span class="p">(</span>
    <span class="n">predictions</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">]],</span>
    <span class="n">rules</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">size</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">minor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Row prediction, but specifically for large deletions</span>

<span class="sd">    Args:</span>
<span class="sd">        predictions ({int: str}): Predictions dictionary</span>
<span class="sd">        rules (pandas.DataFrame): Rules from the catalogue</span>
<span class="sd">        size (float): Percentage of the gene deleted (as a decimal)</span>
<span class="sd">        minor (float): Minors (if existing)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pred</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">current</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
    <span class="c1"># Find which rules are large deletions and see if the mutation is &gt;= rule</span>
    <span class="n">deletion</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;&quot;&quot;del_([01]\.[0-9]+)&quot;&quot;&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">rule</span> <span class="ow">in</span> <span class="n">rules</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">del_match</span> <span class="o">=</span> <span class="n">deletion</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">del_match</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">percentage</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">del_match</span><span class="o">.</span><span class="n">groups</span><span class="p">()[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">size</span> <span class="o">&gt;=</span> <span class="n">percentage</span> <span class="ow">and</span> <span class="n">percentage</span> <span class="o">&gt;</span> <span class="n">current</span><span class="p">:</span>
                <span class="c1"># Match!</span>
                <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                    <span class="c1"># Neither are minor populations so act normally</span>
                    <span class="n">pred</span> <span class="o">=</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">]</span>
                    <span class="n">evidence</span> <span class="o">=</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">]</span>
                    <span class="n">current</span> <span class="o">=</span> <span class="n">percentage</span>
                <span class="k">elif</span> <span class="p">(</span>
                    <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                    <span class="ow">and</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span>
                    <span class="ow">and</span> <span class="n">minor</span> <span class="o">&lt;</span> <span class="mi">1</span>
                    <span class="ow">and</span> <span class="nb">float</span><span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">1</span>
                <span class="p">):</span>
                    <span class="c1"># We have FRS</span>
                    <span class="k">if</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="nb">float</span><span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]):</span>
                        <span class="c1"># Match</span>
                        <span class="n">pred</span> <span class="o">=</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">]</span>
                        <span class="n">evidence</span> <span class="o">=</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">]</span>
                        <span class="n">current</span> <span class="o">=</span> <span class="n">percentage</span>

                <span class="k">elif</span> <span class="p">(</span>
                    <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                    <span class="ow">and</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span>
                    <span class="ow">and</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="mi">1</span>
                    <span class="ow">and</span> <span class="nb">float</span><span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">])</span> <span class="o">&gt;=</span> <span class="mi">1</span>
                <span class="p">):</span>
                    <span class="c1"># We have COV</span>
                    <span class="k">if</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="nb">float</span><span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]):</span>
                        <span class="c1"># Match</span>
                        <span class="n">pred</span> <span class="o">=</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">]</span>
                        <span class="n">evidence</span> <span class="o">=</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">]</span>
                        <span class="n">current</span> <span class="o">=</span> <span class="n">percentage</span>
    <span class="k">if</span> <span class="n">pred</span><span class="p">:</span>
        <span class="n">predictions</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">pred</span><span class="p">),</span> <span class="n">evidence</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.match_multi" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">match_multi</span><span class="p">(</span><span class="n">rule</span><span class="p">,</span> <span class="n">mutation</span><span class="p">,</span> <span class="n">catalogue</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Determine if a given mutation matches a given rule.
This takes into account wildcards.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>rule</code>
            </td>
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Rule to check for</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mutation</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mutation to check for</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>catalogue</code>
            </td>
            <td>
                  <code><span title="grammar_GARC1.Catalogue">Catalogue</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Catalogue this rule is from</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>bool</code></td>            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>True if this mutation matches the rule</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">match_multi</span><span class="p">(</span><span class="n">rule</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">mutation</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">catalogue</span><span class="p">:</span> <span class="n">Catalogue</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Determine if a given mutation matches a given rule.</span>
<span class="sd">    This takes into account wildcards.</span>

<span class="sd">    Args:</span>
<span class="sd">        rule (pandas.Series): Rule to check for</span>
<span class="sd">        mutation (str): Mutation to check for</span>
<span class="sd">        catalogue (Catalogue): Catalogue this rule is from</span>

<span class="sd">    Returns:</span>
<span class="sd">        bool: True if this mutation matches the rule</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">mutation</span><span class="p">:</span>
        <span class="c1"># Literal match</span>
        <span class="k">return</span> <span class="kc">True</span>

    <span class="n">rule_mutations</span> <span class="o">=</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">rule_mutations</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">)):</span>
        <span class="c1"># Not same number of mutations in the rule so not a match</span>
        <span class="k">return</span> <span class="kc">False</span>

    <span class="c1"># Bit more tricky here as we need to check that every part of the mutation matches</span>
    <span class="c1">#   a single part of the rule</span>
    <span class="c1"># do this by constructing a dummy catalogue from the multi-rule and checking each</span>
    <span class="c1">#   part of the mutation hits it</span>

    <span class="c1"># First create a catalogue with just the rules in this multi</span>
    <span class="n">dummy_cat</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;EVIDENCE&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="n">ujson</span><span class="o">.</span><span class="n">dumps</span><span class="p">({</span><span class="s2">&quot;Rule hit&quot;</span><span class="p">:</span> <span class="n">idx</span><span class="p">})</span> <span class="k">for</span> <span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">rule_mutations</span><span class="p">)</span>
        <span class="p">],</span>
        <span class="s2">&quot;MUTATION&quot;</span><span class="p">:</span> <span class="n">rule_mutations</span><span class="p">,</span>
        <span class="s2">&quot;DRUG&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;na&quot;</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">rule_mutations</span><span class="p">],</span>
        <span class="s2">&quot;PREDICTION&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;R&quot;</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">rule_mutations</span><span class="p">],</span>
    <span class="p">}</span>
    <span class="n">r</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">dummy_cat</span><span class="p">)</span>
    <span class="n">r</span><span class="p">[</span>
        <span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">,</span> <span class="s2">&quot;MUTATION&quot;</span><span class="p">,</span> <span class="s2">&quot;POSITION&quot;</span><span class="p">,</span> <span class="s2">&quot;MUTATION_AFFECTS&quot;</span><span class="p">,</span> <span class="s2">&quot;MUTATION_TYPE&quot;</span><span class="p">,</span> <span class="s2">&quot;MINOR&quot;</span><span class="p">]</span>
    <span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">split_mutation</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># As this is a named **tuple**, it is immutable so create a new one</span>
    <span class="n">cat</span> <span class="o">=</span> <span class="n">Catalogue</span><span class="p">(</span>
        <span class="n">catalogue</span><span class="o">.</span><span class="n">genbank_reference</span><span class="p">,</span>
        <span class="n">catalogue</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
        <span class="n">catalogue</span><span class="o">.</span><span class="n">version</span><span class="p">,</span>
        <span class="n">catalogue</span><span class="o">.</span><span class="n">grammar</span><span class="p">,</span>
        <span class="s2">&quot;R&quot;</span><span class="p">,</span>  <span class="c1"># Obviously not R but RFUS are needed here</span>
        <span class="p">[</span><span class="s2">&quot;na&quot;</span><span class="p">],</span>
        <span class="n">r</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
        <span class="p">{</span><span class="s2">&quot;na&quot;</span><span class="p">:</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()},</span>
        <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;na&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()},</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">r</span><span class="p">),</span>
        <span class="n">r</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">pred</span> <span class="o">=</span> <span class="n">predict_GARC1</span><span class="p">(</span><span class="n">cat</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pred</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="c1"># pred == &quot;S&quot;:</span>
                <span class="c1"># No results so not matched</span>
                <span class="k">return</span> <span class="kc">False</span>

            <span class="c1"># Pull out the multi-rule index this hit from the evidence</span>
            <span class="n">pred_ev</span> <span class="o">=</span> <span class="n">ujson</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">pred</span><span class="p">[</span><span class="s2">&quot;na&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>

            <span class="n">rule_idx</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pred_ev</span><span class="p">[</span><span class="s2">&quot;Rule hit&quot;</span><span class="p">])</span>
            <span class="n">r</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="n">rule_idx</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="c1"># Rebuild the catalogue object with the updated rule set</span>
            <span class="n">cat</span> <span class="o">=</span> <span class="n">Catalogue</span><span class="p">(</span>
                <span class="n">catalogue</span><span class="o">.</span><span class="n">genbank_reference</span><span class="p">,</span>
                <span class="n">catalogue</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                <span class="n">catalogue</span><span class="o">.</span><span class="n">version</span><span class="p">,</span>
                <span class="n">catalogue</span><span class="o">.</span><span class="n">grammar</span><span class="p">,</span>
                <span class="s2">&quot;R&quot;</span><span class="p">,</span>  <span class="c1"># Obviously not R but RFUS are needed here</span>
                <span class="p">[</span><span class="s2">&quot;na&quot;</span><span class="p">],</span>
                <span class="n">r</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
                <span class="p">{</span><span class="s2">&quot;na&quot;</span><span class="p">:</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()},</span>
                <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;na&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">r</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()},</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">r</span><span class="p">),</span>
                <span class="n">r</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="c1"># No rules for this drugs etc</span>
            <span class="k">return</span> <span class="kc">False</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">r</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Must have matched every rule in the multi</span>
        <span class="k">return</span> <span class="kc">True</span>

    <span class="c1"># Not matched everything, so not a match</span>
    <span class="k">return</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.merge_predictions" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">merge_predictions</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">catalogue</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>When multi-mutations do not have a hit, they are decomposed and individuals are
    tried instead
This merges these predictions based on the prioritisation defined for this catalogue</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>predictions</code>
            </td>
            <td>
                  <code>{str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>{str: str}}): Dictionary mapping mutation -&gt; effects for
each mutation in the multi</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>catalogue</code>
            </td>
            <td>
                  <code><span title="collections.namedtuple">namedtuple</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The catalogue object
(for finding the values)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    {str: str}: Merged effects dict mapping drug name -&gt; prediction. Or "S" if
        no matches</p>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">merge_predictions</span><span class="p">(</span>
    <span class="n">predictions</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">]</span> <span class="o">|</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span> <span class="n">catalogue</span><span class="p">:</span> <span class="n">Catalogue</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">]</span> <span class="o">|</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;When multi-mutations do not have a hit, they are decomposed and individuals are</span>
<span class="sd">        tried instead</span>
<span class="sd">    This merges these predictions based on the prioritisation defined for this catalogue</span>

<span class="sd">    Args:</span>
<span class="sd">        predictions ({str: {str: str}}): Dictionary mapping mutation -&gt; effects for</span>
<span class="sd">            each mutation in the multi</span>
<span class="sd">        catalogue (collections.namedtuple): The catalogue object</span>
<span class="sd">            (for finding the values)</span>
<span class="sd">    Returns:</span>
<span class="sd">        {str: str}: Merged effects dict mapping drug name -&gt; prediction. Or &quot;S&quot; if</span>
<span class="sd">            no matches</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Pull out all of the drugs which have predictions</span>
    <span class="n">drugs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">pred</span> <span class="ow">in</span> <span class="n">predictions</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">predictions</span><span class="p">[</span><span class="n">pred</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">drugs</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">drug</span><span class="p">)</span>

    <span class="c1"># Default to &#39;S&#39; for now</span>
    <span class="n">merged</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">drug</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="nb">dict</span><span class="p">())</span> <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">drugs</span><span class="p">}</span>

    <span class="c1"># Look for all predictions for each drug, and report the most significant</span>
    <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">drugs</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">predictions</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="c1"># Get the prediction for this drug (if exists)</span>
            <span class="n">this_pred</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="n">mutation</span><span class="p">][</span><span class="n">drug</span><span class="p">]</span>

            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">this_pred</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">pred</span><span class="p">,</span> <span class="n">evidence</span> <span class="o">=</span> <span class="n">this_pred</span>

            <span class="k">if</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">pred</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">merged</span><span class="p">[</span><span class="n">drug</span><span class="p">][</span><span class="mi">0</span><span class="p">]):</span>
                <span class="c1"># New highest priority, so set the prediction</span>
                <span class="n">merged</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">pred</span><span class="p">,</span> <span class="n">evidence</span><span class="p">)</span>

    <span class="c1"># If we have &gt;=1 non-&#39;S&#39; prediction, return the dict</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">([</span><span class="n">key</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">merged</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">merged</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;S&quot;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">drug</span><span class="p">:</span> <span class="n">merged</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">merged</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">merged</span><span class="p">[</span><span class="n">drug</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;S&quot;</span><span class="p">}</span>

    <span class="c1"># Else give the default &#39;S&#39;</span>
    <span class="k">return</span> <span class="s2">&quot;S&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.parse_mutation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_mutation</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Take a GENE_MUTATION, and determine what type it is, what it affects etc</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mutation</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>in the gene_mutation form as defined by GARC e.g. rpoB@S450L or
rpoB@1300_ins_3</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>position</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the position of the mutation. This is the amino acid number in
a protein, or the number of the nucleotide in the promoter, or a
wildcard like '<em>' or '-</em>'</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>mutation_affects</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>whether this mutation affects the promoter (PROM) or
coding sequence (CDS)</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>mutation_type</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>is it a SNP or an INDEL?</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>indel_type</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>indel, ins, del or fs</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>indel_length</code></td>            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>if sufficient information is given, the number of bases
in the INDEL</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>indel_bases</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>if sufficient information is given, the bases in the INDEL</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>before</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the REF amino acid</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
<td><code>after</code></td>            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>the ALT amino acid</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span>
<span class="normal">1228</span>
<span class="normal">1229</span>
<span class="normal">1230</span>
<span class="normal">1231</span>
<span class="normal">1232</span>
<span class="normal">1233</span>
<span class="normal">1234</span>
<span class="normal">1235</span>
<span class="normal">1236</span>
<span class="normal">1237</span>
<span class="normal">1238</span>
<span class="normal">1239</span>
<span class="normal">1240</span>
<span class="normal">1241</span>
<span class="normal">1242</span>
<span class="normal">1243</span>
<span class="normal">1244</span>
<span class="normal">1245</span>
<span class="normal">1246</span>
<span class="normal">1247</span>
<span class="normal">1248</span>
<span class="normal">1249</span>
<span class="normal">1250</span>
<span class="normal">1251</span>
<span class="normal">1252</span>
<span class="normal">1253</span>
<span class="normal">1254</span>
<span class="normal">1255</span>
<span class="normal">1256</span>
<span class="normal">1257</span>
<span class="normal">1258</span>
<span class="normal">1259</span>
<span class="normal">1260</span>
<span class="normal">1261</span>
<span class="normal">1262</span>
<span class="normal">1263</span>
<span class="normal">1264</span>
<span class="normal">1265</span>
<span class="normal">1266</span>
<span class="normal">1267</span>
<span class="normal">1268</span>
<span class="normal">1269</span>
<span class="normal">1270</span>
<span class="normal">1271</span>
<span class="normal">1272</span>
<span class="normal">1273</span>
<span class="normal">1274</span>
<span class="normal">1275</span>
<span class="normal">1276</span>
<span class="normal">1277</span>
<span class="normal">1278</span>
<span class="normal">1279</span>
<span class="normal">1280</span>
<span class="normal">1281</span>
<span class="normal">1282</span>
<span class="normal">1283</span>
<span class="normal">1284</span>
<span class="normal">1285</span>
<span class="normal">1286</span>
<span class="normal">1287</span>
<span class="normal">1288</span>
<span class="normal">1289</span>
<span class="normal">1290</span>
<span class="normal">1291</span>
<span class="normal">1292</span>
<span class="normal">1293</span>
<span class="normal">1294</span>
<span class="normal">1295</span>
<span class="normal">1296</span>
<span class="normal">1297</span>
<span class="normal">1298</span>
<span class="normal">1299</span>
<span class="normal">1300</span>
<span class="normal">1301</span>
<span class="normal">1302</span>
<span class="normal">1303</span>
<span class="normal">1304</span>
<span class="normal">1305</span>
<span class="normal">1306</span>
<span class="normal">1307</span>
<span class="normal">1308</span>
<span class="normal">1309</span>
<span class="normal">1310</span>
<span class="normal">1311</span>
<span class="normal">1312</span>
<span class="normal">1313</span>
<span class="normal">1314</span>
<span class="normal">1315</span>
<span class="normal">1316</span>
<span class="normal">1317</span>
<span class="normal">1318</span>
<span class="normal">1319</span>
<span class="normal">1320</span>
<span class="normal">1321</span>
<span class="normal">1322</span>
<span class="normal">1323</span>
<span class="normal">1324</span>
<span class="normal">1325</span>
<span class="normal">1326</span>
<span class="normal">1327</span>
<span class="normal">1328</span>
<span class="normal">1329</span>
<span class="normal">1330</span>
<span class="normal">1331</span>
<span class="normal">1332</span>
<span class="normal">1333</span>
<span class="normal">1334</span>
<span class="normal">1335</span>
<span class="normal">1336</span>
<span class="normal">1337</span>
<span class="normal">1338</span>
<span class="normal">1339</span>
<span class="normal">1340</span>
<span class="normal">1341</span>
<span class="normal">1342</span>
<span class="normal">1343</span>
<span class="normal">1344</span>
<span class="normal">1345</span>
<span class="normal">1346</span>
<span class="normal">1347</span>
<span class="normal">1348</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">parse_mutation</span><span class="p">(</span>
    <span class="n">mutation</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span>
    <span class="n">Tuple</span><span class="p">[</span>
        <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
        <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
        <span class="nb">str</span><span class="p">,</span>
        <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
        <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
        <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
        <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
        <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">]</span>
<span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Take a GENE_MUTATION, and determine what type it is, what it affects etc</span>

<span class="sd">    Args:</span>
<span class="sd">        mutation (str): in the gene_mutation form as defined by GARC e.g. rpoB@S450L or</span>
<span class="sd">            rpoB@1300_ins_3</span>

<span class="sd">    Returns:</span>
<span class="sd">        position (int): the position of the mutation. This is the amino acid number in</span>
<span class="sd">            a protein, or the number of the nucleotide in the promoter, or a</span>
<span class="sd">            wildcard like &#39;*&#39; or &#39;-*&#39;</span>
<span class="sd">        mutation_affects (str): whether this mutation affects the promoter (PROM) or</span>
<span class="sd">            coding sequence (CDS)</span>
<span class="sd">        mutation_type (str): is it a SNP or an INDEL?</span>
<span class="sd">        indel_type (str): indel, ins, del or fs</span>
<span class="sd">        indel_length (float): if sufficient information is given, the number of bases</span>
<span class="sd">            in the INDEL</span>
<span class="sd">        indel_bases (str): if sufficient information is given, the bases in the INDEL</span>
<span class="sd">        before (str): the REF amino acid</span>
<span class="sd">        after (str): the ALT amino acid</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># set default values</span>
    <span class="n">mutation_type</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">mutation_affects</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">position</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">indel_type</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">indel_length</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">indel_bases</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">before</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">after</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># split the mutation on underscore</span>
    <span class="n">cols</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>

    <span class="c1"># infer what type it is depending on how many elements it contains</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">mutation_type</span> <span class="o">=</span> <span class="s2">&quot;SNP&quot;</span>

        <span class="n">position</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">mutation</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="n">mutation_affects</span> <span class="o">=</span> <span class="n">infer_mutation_affects</span><span class="p">(</span><span class="n">position</span><span class="p">)</span>

        <span class="n">before</span> <span class="o">=</span> <span class="n">mutation</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">after</span> <span class="o">=</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">]:</span>
        <span class="n">mutation_type</span> <span class="o">=</span> <span class="s2">&quot;INDEL&quot;</span>
        <span class="c1"># Checking for large deletions</span>
        <span class="n">percentage</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;del&quot;</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="nb">float</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">percentage</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="n">percentage</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="k">if</span> <span class="n">percentage</span><span class="p">:</span>
            <span class="n">mutation_affects</span> <span class="o">=</span> <span class="s2">&quot;GENE&quot;</span>
            <span class="n">indel_type</span> <span class="o">=</span> <span class="s2">&quot;del&quot;</span>
            <span class="n">indel_length</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">position</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="k">assert</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;Invalid mutation: &quot;</span> <span class="o">+</span> <span class="n">mutation</span>
            <span class="n">mutation_affects</span> <span class="o">=</span> <span class="n">infer_mutation_affects</span><span class="p">(</span><span class="n">position</span><span class="p">)</span>

            <span class="c1"># the third element is one of indel, ins, del or the special case fs</span>
            <span class="n">indel_type</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

            <span class="k">assert</span> <span class="n">indel_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;indel&quot;</span><span class="p">,</span> <span class="s2">&quot;ins&quot;</span><span class="p">,</span> <span class="s2">&quot;del&quot;</span><span class="p">,</span> <span class="s2">&quot;fs&quot;</span><span class="p">,</span> <span class="s2">&quot;mixed&quot;</span><span class="p">],</span> <span class="p">(</span>
                <span class="s2">&quot;form of indel not recognised: &quot;</span> <span class="o">+</span> <span class="n">indel_type</span>
            <span class="p">)</span>

            <span class="c1"># if there is a fourth and final element to an INDEL it is either _4 or</span>
            <span class="c1">#    _ctgc</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
                <span class="k">assert</span> <span class="n">indel_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ins&quot;</span><span class="p">,</span> <span class="s2">&quot;del&quot;</span><span class="p">],</span> <span class="p">(</span>
                    <span class="s2">&quot;form of indel does not make sense when length &quot;</span>
                    <span class="o">+</span> <span class="s2">&quot;or bases specified!: &quot;</span>
                    <span class="o">+</span> <span class="n">indel_type</span>
                <span class="p">)</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">indel_length</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="n">indel_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
                    <span class="n">indel_bases</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
                    <span class="k">assert</span> <span class="mi">0</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span>
                        <span class="n">c</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;a&quot;</span><span class="p">,</span> <span class="s2">&quot;t&quot;</span><span class="p">,</span> <span class="s2">&quot;c&quot;</span><span class="p">,</span> <span class="s2">&quot;g&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">indel_bases</span>
                    <span class="p">],</span> <span class="s2">&quot;only nucleotides of a,t,c,g,z,x are allowed!&quot;</span>

                <span class="k">assert</span> <span class="n">indel_length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="p">(</span>
                    <span class="s2">&quot;makes no sense to have a negative indel length! &quot;</span> <span class="o">+</span> <span class="n">cols</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
                <span class="p">)</span>

    <span class="k">assert</span> <span class="n">mutation_type</span> <span class="ow">in</span> <span class="p">[</span>
        <span class="s2">&quot;INDEL&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SNP&quot;</span><span class="p">,</span>
    <span class="p">],</span> <span class="s2">&quot;form of mutation_type not recognised: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">mutation_type</span><span class="p">)</span>

    <span class="c1"># insist we&#39;ve been given an amino acid or a wildcard only</span>
    <span class="k">if</span> <span class="n">mutation_type</span> <span class="o">==</span> <span class="s2">&quot;SNP&quot;</span><span class="p">:</span>
        <span class="n">sanity_check_snp</span><span class="p">(</span><span class="n">before</span><span class="p">,</span> <span class="n">after</span><span class="p">)</span>

    <span class="n">parsed</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span>
            <span class="n">position</span><span class="p">,</span>
            <span class="n">mutation_affects</span><span class="p">,</span>
            <span class="n">mutation_type</span><span class="p">,</span>
            <span class="n">indel_type</span><span class="p">,</span>
            <span class="n">indel_length</span><span class="p">,</span>
            <span class="n">indel_bases</span><span class="p">,</span>
            <span class="n">before</span><span class="p">,</span>
            <span class="n">after</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">]</span>
    <span class="k">if</span> <span class="p">(</span>
        <span class="n">mutation_affects</span> <span class="o">==</span> <span class="s2">&quot;PROM&quot;</span>
        <span class="ow">and</span> <span class="n">indel_type</span> <span class="o">==</span> <span class="s2">&quot;del&quot;</span>
        <span class="ow">and</span> <span class="n">indel_length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
        <span class="ow">and</span> <span class="n">position</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
    <span class="p">):</span>
        <span class="c1"># Potential case of deletion crossing into coding region</span>
        <span class="k">if</span> <span class="n">position</span> <span class="o">+</span> <span class="n">indel_length</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">new_bases</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">indel_bases</span><span class="p">[</span><span class="nb">abs</span><span class="p">(</span><span class="n">position</span><span class="p">)</span> <span class="p">:]</span> <span class="k">if</span> <span class="n">indel_bases</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="p">)</span>
            <span class="n">parsed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="p">(</span>
                    <span class="mi">1</span><span class="p">,</span>
                    <span class="s2">&quot;CDS&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;INDEL&quot;</span><span class="p">,</span>
                    <span class="s2">&quot;del&quot;</span><span class="p">,</span>
                    <span class="n">position</span> <span class="o">+</span> <span class="n">indel_length</span><span class="p">,</span>
                    <span class="n">new_bases</span><span class="p">,</span>
                    <span class="n">before</span><span class="p">,</span>
                    <span class="n">after</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
    <span class="k">return</span> <span class="n">parsed</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.predict" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">predict</span><span class="p">(</span><span class="n">catalogue</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">mutation</span><span class="p">,</span> <span class="n">gene_mutation</span><span class="p">,</span> <span class="n">drugs</span><span class="p">,</span> <span class="n">result</span><span class="p">,</span> <span class="n">show_evidence</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">mutation_affects</span><span class="p">,</span> <span class="n">mutation_type</span><span class="p">,</span> <span class="n">indel_type</span><span class="p">,</span> <span class="n">indel_length</span><span class="p">,</span> <span class="n">indel_bases</span><span class="p">,</span> <span class="n">before</span><span class="p">,</span> <span class="n">after</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Given a parsed mutation, predict effects for it.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>catalogue</code>
            </td>
            <td>
                  <code><span title="grammar_GARC1.Catalogue">Catalogue</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Catalogue to predict from</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Gene name</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mutation</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mutation in GARC</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_mutation</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Joined gene and mutation (for error messages)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>drugs</code>
            </td>
            <td>
                  <code><span title="typing.List">List</span>[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List of drugs to check for</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>result</code>
            </td>
            <td>
                  <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Tuple">Tuple</span>]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Cumulative results of predictions</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>show_evidence</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to show predictions</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>minor</code>
            </td>
            <td>
                  <code>float | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Minor population supporting this mutation</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>position</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Position this mutation affects</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mutation_affects</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Where this mutation affects (CDS or PROM)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mutation_type</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Type of mutations (INDEL or SNP)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>indel_type</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Type of indel (INS or DEL)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>indel_length</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Length of indel (if any)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>indel_bases</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Bases inserted or deleted (if any)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>before</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Before SNP (if any)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>after</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>After SNP (if any)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Tuple">Tuple</span>] | str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dict[str, Tuple] | Dict[str, str]: Results of predictions</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">predict</span><span class="p">(</span>
    <span class="n">catalogue</span><span class="p">:</span> <span class="n">Catalogue</span><span class="p">,</span>
    <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">mutation</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">gene_mutation</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">drugs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">result</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">],</span>
    <span class="n">show_evidence</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">minor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">position</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">mutation_affects</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">mutation_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">indel_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">indel_length</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">indel_bases</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">before</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">after</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">]</span> <span class="o">|</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given a parsed mutation, predict effects for it.</span>

<span class="sd">    Args:</span>
<span class="sd">        catalogue (Catalogue): Catalogue to predict from</span>
<span class="sd">        gene (str): Gene name</span>
<span class="sd">        mutation (str): Mutation in GARC</span>
<span class="sd">        gene_mutation (str): Joined gene and mutation (for error messages)</span>
<span class="sd">        drugs (List[str]): List of drugs to check for</span>
<span class="sd">        result (Dict[str, Tuple]): Cumulative results of predictions</span>
<span class="sd">        show_evidence (bool): Whether to show predictions</span>
<span class="sd">        minor (float | None): Minor population supporting this mutation</span>
<span class="sd">        position (int): Position this mutation affects</span>
<span class="sd">        mutation_affects (str): Where this mutation affects (CDS or PROM)</span>
<span class="sd">        mutation_type (str): Type of mutations (INDEL or SNP)</span>
<span class="sd">        indel_type (str): Type of indel (INS or DEL)</span>
<span class="sd">        indel_length (int): Length of indel (if any)</span>
<span class="sd">        indel_bases (str): Bases inserted or deleted (if any)</span>
<span class="sd">        before (str): Before SNP (if any)</span>
<span class="sd">        after (str): After SNP (if any)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Dict[str, Tuple] | Dict[str, str]: Results of predictions</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># create the vectors of Booleans that will apply</span>
    <span class="n">position_vector</span> <span class="o">=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="o">.</span><span class="n">POSITION</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
        <span class="p">[</span><span class="n">position</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">position</span><span class="p">),</span> <span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="s2">&quot;-*&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">mutation_affects_vector</span> <span class="o">=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION_AFFECTS</span> <span class="o">==</span> <span class="n">mutation_affects</span>
    <span class="n">mutation_type_vector</span> <span class="o">=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION_TYPE</span> <span class="o">==</span> <span class="n">mutation_type</span>
    <span class="n">gene_vector</span> <span class="o">=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="o">.</span><span class="n">GENE</span> <span class="o">==</span> <span class="n">gene</span>
    <span class="n">subset_vector</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">position_vector</span> <span class="o">&amp;</span> <span class="n">mutation_affects_vector</span> <span class="o">&amp;</span> <span class="n">mutation_type_vector</span> <span class="o">&amp;</span> <span class="n">gene_vector</span>
    <span class="p">)</span>
    <span class="c1"># deal with each compound, one at a time</span>
    <span class="k">for</span> <span class="n">compound</span> <span class="ow">in</span> <span class="n">drugs</span><span class="p">:</span>
        <span class="n">subset_rules</span> <span class="o">=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="p">(</span><span class="n">subset_vector</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="o">.</span><span class="n">DRUG</span> <span class="o">==</span> <span class="n">compound</span><span class="p">)</span>
        <span class="p">]</span>

        <span class="c1"># prepare a dictionary to store hits with the priority as the key:</span>
        <span class="c1">#   e.g. {10:&#39;R&#39;,5:&#39;U&#39;}</span>
        <span class="n">predictions</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">]]</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">subset_rules</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">subset_position_vector</span> <span class="o">=</span> <span class="n">subset_rules</span><span class="o">.</span><span class="n">POSITION</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">position</span><span class="p">)</span>
            <span class="n">subset_mutation_type_vector</span> <span class="o">=</span> <span class="n">subset_rules</span><span class="o">.</span><span class="n">MUTATION_TYPE</span> <span class="o">==</span> <span class="n">mutation_type</span>

            <span class="k">if</span> <span class="n">mutation_type</span> <span class="o">==</span> <span class="s2">&quot;SNP&quot;</span><span class="p">:</span>
                <span class="n">process_snp_variants</span><span class="p">(</span>
                    <span class="n">mutation_affects</span><span class="p">,</span>
                    <span class="n">predictions</span><span class="p">,</span>
                    <span class="n">before</span><span class="p">,</span>
                    <span class="n">after</span><span class="p">,</span>
                    <span class="n">mutation</span><span class="p">,</span>
                    <span class="n">subset_rules</span><span class="p">,</span>
                    <span class="n">subset_mutation_type_vector</span><span class="p">,</span>
                    <span class="n">subset_position_vector</span><span class="p">,</span>
                    <span class="n">minor</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="k">elif</span> <span class="n">mutation_type</span> <span class="o">==</span> <span class="s2">&quot;INDEL&quot;</span><span class="p">:</span>
                <span class="n">process_indel_variants</span><span class="p">(</span>
                    <span class="n">mutation_affects</span><span class="p">,</span>
                    <span class="n">predictions</span><span class="p">,</span>
                    <span class="n">subset_rules</span><span class="p">,</span>
                    <span class="n">subset_mutation_type_vector</span><span class="p">,</span>
                    <span class="n">subset_position_vector</span><span class="p">,</span>
                    <span class="n">indel_length</span><span class="p">,</span>
                    <span class="n">indel_type</span><span class="p">,</span>
                    <span class="n">indel_bases</span><span class="p">,</span>
                    <span class="n">position</span><span class="p">,</span>
                    <span class="n">minor</span><span class="p">,</span>
                <span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">predictions</span><span class="p">:</span>
            <span class="c1"># all mutations should hit at least one of the default entries, so if this</span>
            <span class="c1">#   doesn&#39;t happen, something is wrong UNLESS the mutation given is a minor allele</span>
            <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">gene_mutation</span><span class="p">:</span>
                <span class="n">result</span><span class="p">[</span><span class="n">compound</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="p">{})</span>
            <span class="k">elif</span> <span class="n">mutation_type</span> <span class="o">==</span> <span class="s2">&quot;SNP&quot;</span> <span class="ow">and</span> <span class="n">after</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">]:</span>
                <span class="c1"># Null call with no specific row matches, so return S</span>
                <span class="c1"># as it doesn&#39;t make sense to match defaults for a null call</span>
                <span class="n">result</span><span class="p">[</span><span class="n">compound</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="p">{})</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;No entry found in the catalogue for &quot;</span>
                    <span class="o">+</span> <span class="n">gene_mutation</span>
                    <span class="o">+</span> <span class="s2">&quot; &quot;</span>
                    <span class="o">+</span> <span class="n">compound</span>
                <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">final_prediction</span><span class="p">:</span> <span class="n">Tuple</span> <span class="o">=</span> <span class="n">predictions</span><span class="p">[</span><span class="nb">sorted</span><span class="p">(</span><span class="n">predictions</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span>
            <span class="n">result</span><span class="p">[</span><span class="n">compound</span><span class="p">]</span> <span class="o">=</span> <span class="n">final_prediction</span>

    <span class="k">return</span> <span class="n">result</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.predict_GARC1" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">predict_GARC1</span><span class="p">(</span><span class="n">catalogue</span><span class="p">,</span> <span class="n">gene_mutation</span><span class="p">,</span> <span class="n">show_evidence</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>For the GARC1 grammar, predict the effect of the supplied gene_mutation.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>catalogue</code>
            </td>
            <td>
                  <code><span title="collections.namedtuple">namedtuple</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>defines the resistance catalogue</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_mutation</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>as specified by the GARC1 grammar, e.g. rpoB@S450L,
fabG1@c-15t, ahpC@128_indel</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>show_evidence</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Flag for whether to return the evidence for
this prediction</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Dict">Dict</span>[str, <span title="typing.Tuple">Tuple</span>] | <span title="typing.Dict">Dict</span>[str, str] | str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>{str: (str, str) | str} | str: Either a dictionary mapping drug name -&gt;
(prediction, evidence) if show_evidence is True or "S" if no hits
in the catalogue</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">predict_GARC1</span><span class="p">(</span>
    <span class="n">catalogue</span><span class="p">:</span> <span class="n">Catalogue</span><span class="p">,</span> <span class="n">gene_mutation</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">show_evidence</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">]</span> <span class="o">|</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    For the GARC1 grammar, predict the effect of the supplied gene_mutation.</span>

<span class="sd">    Args:</span>
<span class="sd">        catalogue (collections.namedtuple): defines the resistance catalogue</span>
<span class="sd">        gene_mutation (str): as specified by the GARC1 grammar, e.g. rpoB@S450L,</span>
<span class="sd">            fabG1@c-15t, ahpC@128_indel</span>
<span class="sd">        show_evidence (bool, optional): Flag for whether to return the evidence for</span>
<span class="sd">            this prediction</span>

<span class="sd">    Returns:</span>
<span class="sd">        {str: (str, str) | str} | str: Either a dictionary mapping drug name -&gt;</span>
<span class="sd">            (prediction, evidence) if show_evidence is True or &quot;S&quot; if no hits</span>
<span class="sd">            in the catalogue</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># create the result dictionary e.g. {&quot;RIF&quot;:&#39;R&#39;, &quot;RFB&quot;:&#39;R&#39;}</span>
    <span class="c1"># most of the time the predictions will be identical, but this allows them to</span>
    <span class="c1">#   diverge in future</span>
    <span class="n">result</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">if</span> <span class="s2">&quot;&amp;&quot;</span> <span class="ow">in</span> <span class="n">gene_mutation</span><span class="p">:</span>
        <span class="c1"># Multi-gene so treat differently</span>
        <span class="n">result_multi</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">]</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">predict_multi</span><span class="p">(</span><span class="n">catalogue</span><span class="p">,</span> <span class="n">gene_mutation</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">show_evidence</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">result_multi</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">result_multi</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">result_multi</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">result_multi</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>

    <span class="n">components</span> <span class="o">=</span> <span class="n">gene_mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;@&quot;</span><span class="p">)</span>

    <span class="n">gene</span> <span class="o">=</span> <span class="n">components</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">mutation</span> <span class="o">=</span> <span class="n">gene_mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">gene</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Check for minors</span>
    <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">minor</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">assert</span> <span class="kc">False</span><span class="p">,</span> <span class="s2">&quot;Malformed mutation! &quot;</span> <span class="o">+</span> <span class="n">mutation</span>
        <span class="n">mutation</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">assert</span> <span class="n">minor</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="s2">&quot;Minor population given with no evidence! &quot;</span> <span class="o">+</span> <span class="n">mutation</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">minor</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># parse the mutation to work out what type of mutation it is, where it acts etc</span>
    <span class="n">parsed_mutations</span> <span class="o">=</span> <span class="n">parse_mutation</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span>

    <span class="c1"># if the gene isn&#39;t in the catalogue, then assume it has no effect</span>
    <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">genes</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;S&quot;</span>

    <span class="c1"># find out the drugs to which changes in this gene can confer resistance</span>
    <span class="n">drugs</span> <span class="o">=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">gene_lookup</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span>

    <span class="k">for</span> <span class="p">(</span>
        <span class="n">position</span><span class="p">,</span>
        <span class="n">mutation_affects</span><span class="p">,</span>
        <span class="n">mutation_type</span><span class="p">,</span>
        <span class="n">indel_type</span><span class="p">,</span>
        <span class="n">indel_length</span><span class="p">,</span>
        <span class="n">indel_bases</span><span class="p">,</span>
        <span class="n">before</span><span class="p">,</span>
        <span class="n">after</span><span class="p">,</span>
    <span class="p">)</span> <span class="ow">in</span> <span class="n">parsed_mutations</span><span class="p">:</span>
        <span class="n">res</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span>
            <span class="n">catalogue</span><span class="p">,</span>
            <span class="n">gene</span><span class="p">,</span>
            <span class="n">mutation</span><span class="p">,</span>
            <span class="n">gene_mutation</span><span class="p">,</span>
            <span class="n">drugs</span><span class="p">,</span>
            <span class="n">result</span><span class="p">,</span>
            <span class="n">show_evidence</span><span class="p">,</span>
            <span class="n">minor</span><span class="p">,</span>
            <span class="n">position</span><span class="p">,</span>
            <span class="n">mutation_affects</span><span class="p">,</span>
            <span class="n">mutation_type</span><span class="p">,</span>
            <span class="n">indel_type</span><span class="p">,</span>
            <span class="n">indel_length</span><span class="p">,</span>
            <span class="n">indel_bases</span><span class="p">,</span>
            <span class="n">before</span><span class="p">,</span>
            <span class="n">after</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">res</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">res</span>

    <span class="c1"># Null calls need a little nudge to ensure that they are correctly handled if they don&#39;t hit any rules</span>
    <span class="k">if</span> <span class="n">mutation_type</span> <span class="o">==</span> <span class="s2">&quot;SNP&quot;</span> <span class="ow">and</span> <span class="n">after</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">]:</span>
        <span class="n">all_default_nulls</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">drugs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">result</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">!=</span> <span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="p">{}):</span>
                <span class="n">all_default_nulls</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="n">all_default_nulls</span><span class="p">:</span>
            <span class="k">return</span> <span class="s2">&quot;S&quot;</span>

    <span class="k">if</span> <span class="n">show_evidence</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">result</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">result</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">result</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>

    <span class="k">return</span> <span class="n">result</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.predict_multi" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">predict_multi</span><span class="p">(</span><span class="n">catalogue</span><span class="p">,</span> <span class="n">gene_mutation</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Get the predictions for a given multi-mutation.
Mutli-mutations are (currently) a lot stricter than others, and do not support
    wildcards or subsets of the mutations</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>catalogue</code>
            </td>
            <td>
                  <code><span title="collections.namedtuple">namedtuple</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The resistance catalogue</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_mutation</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>String of the mutations, each in GARC, separated by <code>&amp;</code></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    {str: str} | str : Either a dict with drugs as the keys, or 'S' when no
        predictions in catalogue</p>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">predict_multi</span><span class="p">(</span><span class="n">catalogue</span><span class="p">:</span> <span class="n">Catalogue</span><span class="p">,</span> <span class="n">gene_mutation</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">]</span> <span class="o">|</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the predictions for a given multi-mutation.</span>
<span class="sd">    Mutli-mutations are (currently) a lot stricter than others, and do not support</span>
<span class="sd">        wildcards or subsets of the mutations</span>

<span class="sd">    Args:</span>
<span class="sd">        catalogue (collections.namedtuple): The resistance catalogue</span>
<span class="sd">        gene_mutation (str): String of the mutations, each in GARC, separated by `&amp;`</span>
<span class="sd">    Returns:</span>
<span class="sd">        {str: str} | str : Either a dict with drugs as the keys, or &#39;S&#39; when no</span>
<span class="sd">            predictions in catalogue</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Ensure that the mutations are in a reproducable order</span>
    <span class="n">sorted_mutation</span> <span class="o">=</span> <span class="s2">&quot;&amp;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">gene_mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">))))</span>

    <span class="c1"># Check for minor populations (and remove, we&#39;ll check these separately)</span>
    <span class="n">minors</span> <span class="o">=</span> <span class="p">[</span>
        <span class="nb">float</span><span class="p">(</span><span class="n">mut</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">mut</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">sorted_mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="n">sorted_mutation</span> <span class="o">=</span> <span class="s2">&quot;&amp;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="p">[</span><span class="n">mut</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">sorted_mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">)]</span>
    <span class="p">)</span>

    <span class="c1"># Check epistasis first</span>
    <span class="n">epi_rules</span> <span class="o">=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="p">[</span><span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="p">[</span><span class="s2">&quot;MUTATION_TYPE&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;EPISTASIS&quot;</span><span class="p">]</span>
    <span class="c1"># Note the use of &quot;&quot; here instead of &quot;{}&quot; to allow empty evidence fields</span>
    <span class="n">epi_drugs</span> <span class="o">=</span> <span class="p">{</span><span class="n">drug</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">drugs</span><span class="p">}</span>
    <span class="n">values</span> <span class="o">=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">values</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">rule</span> <span class="ow">in</span> <span class="n">epi_rules</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">match_multi</span><span class="p">(</span><span class="n">rule</span><span class="p">,</span> <span class="n">sorted_mutation</span><span class="p">,</span> <span class="n">catalogue</span><span class="p">):</span>
            <span class="c1"># Epistasis rule hit</span>
            <span class="n">drug</span> <span class="o">=</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;DRUG&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">epi_drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="c1"># This drug already has an epistasis rule hit so throw an error</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Conflicting epistasis rules for </span><span class="si">{</span><span class="n">gene_mutation</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">drug</span><span class="si">}</span><span class="s2">! &quot;</span>
                    <span class="s2">&quot;Check your catalogue!&quot;</span>
                <span class="p">)</span>
            <span class="c1"># Valid, so check for minors</span>
            <span class="k">for</span> <span class="n">cat</span><span class="p">,</span> <span class="n">minor</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">),</span> <span class="n">minors</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">cat</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                    <span class="c1"># Neither are minors so add</span>
                    <span class="n">epi_drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">],</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">])</span>
                <span class="k">elif</span> <span class="n">cat</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">minor</span> <span class="o">&lt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="nb">float</span><span class="p">(</span><span class="n">cat</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="c1"># FRS</span>
                    <span class="k">if</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="nb">float</span><span class="p">(</span><span class="n">cat</span><span class="p">):</span>
                        <span class="c1"># Match</span>
                        <span class="n">epi_drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">],</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">])</span>
                <span class="k">elif</span> <span class="n">cat</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="mi">1</span> <span class="ow">and</span> <span class="nb">float</span><span class="p">(</span><span class="n">cat</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="c1"># COV</span>
                    <span class="k">if</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="nb">float</span><span class="p">(</span><span class="n">cat</span><span class="p">):</span>
                        <span class="c1"># Match</span>
                        <span class="n">epi_drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">],</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">])</span>

    <span class="c1"># Get the multi rules</span>
    <span class="n">multi_rules</span> <span class="o">=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="p">[</span><span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="p">[</span><span class="s2">&quot;MUTATION_TYPE&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;MULTI&quot;</span><span class="p">]</span>

    <span class="n">drugs</span> <span class="o">=</span> <span class="p">{</span><span class="n">drug</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">drugs</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">rule</span> <span class="ow">in</span> <span class="n">multi_rules</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">drug</span> <span class="o">=</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;DRUG&quot;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">epi_drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
            <span class="c1"># Epistasis rule already covers this, so skip it</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="n">match_multi</span><span class="p">(</span><span class="n">rule</span><span class="p">,</span> <span class="n">sorted_mutation</span><span class="p">,</span> <span class="n">catalogue</span><span class="p">):</span>
            <span class="c1"># We have a match! Prioritise predictions based on values</span>
            <span class="k">if</span> <span class="n">values</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">])</span> <span class="o">&lt;</span> <span class="n">values</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">][</span><span class="mi">0</span><span class="p">]):</span>
                <span class="c1"># The prediction is closer to the start of the values list, so should</span>
                <span class="c1">#   take priority</span>
                <span class="c1"># Check for minor populations first though</span>
                <span class="k">for</span> <span class="n">cat</span><span class="p">,</span> <span class="n">minor</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">),</span> <span class="n">minors</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">cat</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                        <span class="c1"># Neither are minors so add</span>
                        <span class="n">drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">],</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">])</span>
                    <span class="k">elif</span> <span class="p">(</span>
                        <span class="n">cat</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">minor</span> <span class="o">&lt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="nb">float</span><span class="p">(</span><span class="n">cat</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">1</span>
                    <span class="p">):</span>
                        <span class="c1"># FRS</span>
                        <span class="k">if</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="nb">float</span><span class="p">(</span><span class="n">cat</span><span class="p">):</span>
                            <span class="c1"># Match</span>
                            <span class="n">drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">],</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">])</span>
                    <span class="k">elif</span> <span class="p">(</span>
                        <span class="n">cat</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span>
                        <span class="ow">and</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                        <span class="ow">and</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="mi">1</span>
                        <span class="ow">and</span> <span class="nb">float</span><span class="p">(</span><span class="n">cat</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">1</span>
                    <span class="p">):</span>
                        <span class="c1"># COV</span>
                        <span class="k">if</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="nb">float</span><span class="p">(</span><span class="n">cat</span><span class="p">):</span>
                            <span class="c1"># Match</span>
                            <span class="n">drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">rule</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">],</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">])</span>

    <span class="c1"># Check to ensure we have at least 1 prediction</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">([</span><span class="n">key</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">epi_drugs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">epi_drugs</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># Epistasis hit(s) so join with the multis</span>
        <span class="n">to_return</span> <span class="o">=</span> <span class="n">drugs</span>
        <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">epi_drugs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">epi_drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="c1"># This drug has an epi match</span>
                <span class="n">to_return</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">=</span> <span class="n">epi_drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;S&quot;</span><span class="p">:</span>
                <span class="c1"># Drop default `S` predictions</span>
                <span class="k">del</span> <span class="n">to_return</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span>
        <span class="c1"># For whatever reason, this appeases mypy</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">value</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">to_return</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">([</span><span class="n">key</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">drugs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">drugs</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;S&quot;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># At least one multi hit</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">drug</span><span class="p">:</span> <span class="n">drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">drugs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">if</span> <span class="n">drugs</span><span class="p">[</span><span class="n">drug</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;S&quot;</span><span class="p">}</span>

    <span class="c1"># Nothing predicted, so try each individual mutation</span>
    <span class="n">predictions</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">mutation</span><span class="p">,</span> <span class="n">minor</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">sorted_mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">),</span> <span class="n">minors</span><span class="p">):</span>
        <span class="c1"># Put minor populations back into mutations as required</span>
        <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">mutation</span> <span class="o">=</span> <span class="n">mutation</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">minor</span><span class="p">)</span>

        <span class="c1"># Get the prediction for the mutation</span>
        <span class="n">pred</span> <span class="o">=</span> <span class="n">predict_GARC1</span><span class="p">(</span><span class="n">catalogue</span><span class="p">,</span> <span class="n">mutation</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pred</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="c1"># We have a &#39;S&#39; prediction so ignore it</span>
            <span class="k">continue</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">predictions</span><span class="p">[</span><span class="n">mutation</span><span class="p">]</span> <span class="o">=</span> <span class="n">pred</span>

    <span class="k">return</span> <span class="n">merge_predictions</span><span class="p">(</span><span class="n">predictions</span><span class="p">,</span> <span class="n">catalogue</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.process_catalogue_GARC1" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_catalogue_GARC1</span><span class="p">(</span><span class="n">rules</span><span class="p">,</span> <span class="n">drugs</span><span class="p">,</span> <span class="n">catalogue_genes_only</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>For the GARC1 grammar, add some additional columns to the rules dataframe that
    can be inferred from the mutation.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>rules</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pandas dataframe of the AMR catalogue in the
general form</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>drugs</code>
            </td>
            <td>
                  <code>[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list of drugs in the AMR catalogue</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>catalogue_genes_only</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>whether to subset the catalogue down so it ONLY
includes (DRUG,GENE) pairs that include at least one row predicting
resistance</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    (pandas.DataFrame, [str], {str: [str]}, {str: [str]}) : Tuple of values.
        In order:
            DataFrame of the rules
            List of the genes
            Dictionary mapping drug name -&gt; [gene names]
            Dictionary mapping gene name -&gt; [drug names]</p>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_catalogue_GARC1</span><span class="p">(</span>
    <span class="n">rules</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">drugs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">catalogue_genes_only</span><span class="p">:</span> <span class="nb">bool</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]],</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    For the GARC1 grammar, add some additional columns to the rules dataframe that</span>
<span class="sd">        can be inferred from the mutation.</span>

<span class="sd">    Args:</span>
<span class="sd">        rules (pandas.DataFrame): Pandas dataframe of the AMR catalogue in the</span>
<span class="sd">            general form</span>
<span class="sd">        drugs ([str]): list of drugs in the AMR catalogue</span>
<span class="sd">        catalogue_genes_only (bool): whether to subset the catalogue down so it ONLY</span>
<span class="sd">            includes (DRUG,GENE) pairs that include at least one row predicting</span>
<span class="sd">            resistance</span>
<span class="sd">    Returns:</span>
<span class="sd">        (pandas.DataFrame, [str], {str: [str]}, {str: [str]}) : Tuple of values.</span>
<span class="sd">            In order:</span>
<span class="sd">                DataFrame of the rules</span>
<span class="sd">                List of the genes</span>
<span class="sd">                Dictionary mapping drug name -&gt; [gene names]</span>
<span class="sd">                Dictionary mapping gene name -&gt; [drug names]</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># infer these columns</span>
    <span class="n">rules</span><span class="p">[</span>
        <span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">,</span> <span class="s2">&quot;MUTATION&quot;</span><span class="p">,</span> <span class="s2">&quot;POSITION&quot;</span><span class="p">,</span> <span class="s2">&quot;MUTATION_AFFECTS&quot;</span><span class="p">,</span> <span class="s2">&quot;MUTATION_TYPE&quot;</span><span class="p">,</span> <span class="s2">&quot;MINOR&quot;</span><span class="p">]</span>
    <span class="p">]</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">split_mutation</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">catalogue_genes_only</span><span class="p">:</span>
        <span class="c1"># create an index of which (drug,gene) pairs have at least one row predicting</span>
        <span class="c1">#   resistance or uncertainty</span>
        <span class="n">drug_list_index</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">rules</span><span class="o">.</span><span class="n">PREDICTION</span> <span class="o">==</span> <span class="s2">&quot;R&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">PREDICTION</span> <span class="o">==</span> <span class="s2">&quot;U&quot;</span><span class="p">)][</span>
                <span class="p">[</span><span class="s2">&quot;DRUG&quot;</span><span class="p">,</span> <span class="s2">&quot;GENE&quot;</span><span class="p">]</span>
            <span class="p">]</span>
            <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;DRUG&quot;</span><span class="p">,</span> <span class="s2">&quot;GENE&quot;</span><span class="p">])</span>
            <span class="o">.</span><span class="n">count</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="c1"># Exclude multi/epistasis rules from this as epistasis rules override R</span>
        <span class="n">multis</span> <span class="o">=</span> <span class="n">rules</span><span class="p">[</span>
            <span class="p">(</span><span class="n">rules</span><span class="p">[</span><span class="s2">&quot;MUTATION_TYPE&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;EPISTASIS&quot;</span><span class="p">)</span>
            <span class="o">|</span> <span class="p">(</span><span class="n">rules</span><span class="p">[</span><span class="s2">&quot;MUTATION_TYPE&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;MULTI&quot;</span><span class="p">)</span>
        <span class="p">]</span>

        <span class="c1"># index the rules so we can right join</span>
        <span class="n">rules</span><span class="o">.</span><span class="n">set_index</span><span class="p">([</span><span class="s2">&quot;DRUG&quot;</span><span class="p">,</span> <span class="s2">&quot;GENE&quot;</span><span class="p">],</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># do the right join; this has the effect of removing all the rows for</span>
        <span class="c1">#   &quot;genes of interest&quot;</span>
        <span class="n">rules</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">drug_list_index</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;right&quot;</span><span class="p">)</span>

        <span class="c1"># remove the index</span>
        <span class="n">rules</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Add back multi/epistais rules now</span>
        <span class="n">rules</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">rules</span><span class="p">,</span> <span class="n">multis</span><span class="p">])</span>
        <span class="c1"># Some multis (but not all) can be duplicated here, so drop duplicate entries</span>
        <span class="c1"># Note that drop_duplicates doesn&#39;t like the JSON columns, so de-duplicate on other fields</span>
        <span class="n">rules</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span>
            <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">subset</span><span class="o">=</span><span class="p">[</span>
                <span class="s2">&quot;DRUG&quot;</span><span class="p">,</span>
                <span class="s2">&quot;GENE&quot;</span><span class="p">,</span>
                <span class="s2">&quot;MUTATION&quot;</span><span class="p">,</span>
                <span class="s2">&quot;POSITION&quot;</span><span class="p">,</span>
                <span class="s2">&quot;MUTATION_AFFECTS&quot;</span><span class="p">,</span>
                <span class="s2">&quot;MUTATION_TYPE&quot;</span><span class="p">,</span>
                <span class="s2">&quot;MINOR&quot;</span><span class="p">,</span>
            <span class="p">],</span>
        <span class="p">)</span>

    <span class="c1"># create a list of the genes mentioned in the catalogue</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">GENE</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

    <span class="c1"># create a dictionary where the genes are keys and the entries tell which drugs</span>
    <span class="c1">#   are affected</span>
    <span class="n">gene_lookup</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rules</span><span class="o">.</span><span class="n">GENE</span> <span class="o">==</span> <span class="n">gene_name</span><span class="p">]</span>
        <span class="n">gene_lookup</span><span class="p">[</span><span class="n">gene_name</span><span class="p">]</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">DRUG</span><span class="o">.</span><span class="n">unique</span><span class="p">()))</span>

    <span class="c1"># create a dictionary where the drugs are keys and the entries tell which genes</span>
    <span class="c1">#   are affected</span>
    <span class="n">drug_lookup</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">drug_name</span> <span class="ow">in</span> <span class="n">drugs</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rules</span><span class="o">.</span><span class="n">DRUG</span> <span class="o">==</span> <span class="n">drug_name</span><span class="p">]</span>
        <span class="n">drug_lookup</span><span class="p">[</span><span class="n">drug_name</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">GENE</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>

    <span class="k">return</span> <span class="n">rules</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">drug_lookup</span><span class="p">,</span> <span class="n">gene_lookup</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.process_indel_variants" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_indel_variants</span><span class="p">(</span><span class="n">mutation_affects</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="n">rules</span><span class="p">,</span> <span class="n">rules_mutation_type_vector</span><span class="p">,</span> <span class="n">rules_position_vector</span><span class="p">,</span> <span class="n">indel_length</span><span class="p">,</span> <span class="n">indel_type</span><span class="p">,</span> <span class="n">indel_bases</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Apply the cascading rules for INDELs in GARC</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mutation_affects</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Which area this mutation affects. i.e PROM/CDS</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>predictions</code>
            </td>
            <td>
                  <code>{int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str}): Dictionary mapping priority -&gt; prediction. This is
updated for implict return</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rules</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Rules DataFrame</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rules_mutation_type_vector</code>
            </td>
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Series for the rules which refer
to this mutation type</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rules_position_vector</code>
            </td>
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Series for the rules which refer to
this position</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>indel_length</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Length of the indel</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>indel_type</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Type of the indel. i.e ins/del/fs/indel</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>indel_bases</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Bases inserted/deleted (if applicable)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>position</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Position of the indel</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>minor</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Float for supporting evidence of minor populations (or None if
not a minor population)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_indel_variants</span><span class="p">(</span>
    <span class="n">mutation_affects</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">predictions</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">]],</span>
    <span class="n">rules</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">rules_mutation_type_vector</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span>
    <span class="n">rules_position_vector</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span>
    <span class="n">indel_length</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">indel_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">indel_bases</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">position</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">minor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Apply the cascading rules for INDELs in GARC</span>

<span class="sd">    Args:</span>
<span class="sd">        mutation_affects (str): Which area this mutation affects. i.e PROM/CDS</span>
<span class="sd">        predictions ({int: str}): Dictionary mapping priority -&gt; prediction. This is</span>
<span class="sd">            updated for implict return</span>
<span class="sd">        rules (pandas.DataFrame): Rules DataFrame</span>
<span class="sd">        rules_mutation_type_vector (pandas.Series): Series for the rules which refer</span>
<span class="sd">            to this mutation type</span>
<span class="sd">        rules_position_vector (pandas.Series): Series for the rules which refer to</span>
<span class="sd">            this position</span>
<span class="sd">        indel_length (int): Length of the indel</span>
<span class="sd">        indel_type (str): Type of the indel. i.e ins/del/fs/indel</span>
<span class="sd">        indel_bases (str): Bases inserted/deleted (if applicable)</span>
<span class="sd">        position (int): Position of the indel</span>
<span class="sd">        minor (float): Float for supporting evidence of minor populations (or None if</span>
<span class="sd">            not a minor population)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">mutation_affects</span> <span class="o">==</span> <span class="s2">&quot;GENE&quot;</span> <span class="ow">and</span> <span class="n">indel_length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Large deletions are priority 1</span>
        <span class="n">large_del</span><span class="p">(</span>
            <span class="n">predictions</span><span class="p">,</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rules_mutation_type_vector</span><span class="p">],</span> <span class="n">indel_length</span><span class="p">,</span> <span class="n">minor</span>
        <span class="p">)</span>

    <span class="c1"># PRIORITY 1: any insertion or deletion in the CDS or PROM (e.g. rpoB@*_indel</span>
    <span class="c1">#   or rpoB@-*_indel)</span>
    <span class="k">if</span> <span class="n">mutation_affects</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="n">rules_mutation_type_vector</span>
            <span class="o">&amp;</span> <span class="p">((</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;*_indel&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;*_mixed&quot;</span><span class="p">))</span>
        <span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="n">rules_mutation_type_vector</span>
            <span class="o">&amp;</span> <span class="p">((</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;-*_indel&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;-*mixed&quot;</span><span class="p">))</span>
        <span class="p">]</span>
    <span class="c1"># any insertion or deletion in the CDS or PROM</span>
    <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

    <span class="c1"># PRIORITY 2: rpoB@*_ins, rpoB@*_del any insertion (or deletion) in the CDS or PROM</span>
    <span class="k">if</span> <span class="n">indel_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">mutation_affects</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                <span class="n">rules_mutation_type_vector</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;*_&quot;</span> <span class="o">+</span> <span class="n">indel_type</span><span class="p">]))</span>
            <span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                <span class="n">rules_mutation_type_vector</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;-*_&quot;</span> <span class="o">+</span> <span class="n">indel_type</span><span class="p">]))</span>
            <span class="p">]</span>
        <span class="c1"># any insertion (or deletion) in the CDS or PROM</span>
        <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

    <span class="c1"># PRIORITY 3: any insertion of a specified length (or deletion) in the CDS or PROM</span>
    <span class="c1">#   (e.g. rpoB@*_ins_2, rpoB@*_del_3, rpoB@-*_ins_1, rpoB@-*_del_200)</span>
    <span class="k">if</span> <span class="n">indel_length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">indel_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">indel_type</span> <span class="o">!=</span> <span class="s2">&quot;indel&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">mutation_affects</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                <span class="n">rules_mutation_type_vector</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;*_&quot;</span> <span class="o">+</span> <span class="n">indel_type</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">indel_length</span><span class="p">)]))</span>
            <span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                <span class="n">rules_mutation_type_vector</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;-*_&quot;</span> <span class="o">+</span> <span class="n">indel_type</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">indel_length</span><span class="p">)]))</span>
            <span class="p">]</span>
        <span class="c1"># any insertion of a specified length (or deletion) in the CDS or PROM</span>
        <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

    <span class="c1"># PRIORITY=4: any frameshifting insertion or deletion in the CDS (e.g. rpoB@*_fs)</span>
    <span class="k">if</span> <span class="n">indel_length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="p">(</span><span class="n">indel_length</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rules_mutation_type_vector</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;*_fs&quot;</span><span class="p">)]</span>
        <span class="c1"># any frameshifting insertion or deletion in the CDS</span>
        <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

    <span class="c1"># PRIORITY=5: any indel at a specific position in the CDS or PROM</span>
    <span class="c1">#   (e.g. rpoB@1300_indel or rpoB@-15_indel)</span>
    <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
        <span class="n">rules_mutation_type_vector</span>
        <span class="o">&amp;</span> <span class="n">rules_position_vector</span>
        <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;indel&quot;</span><span class="p">))</span>
    <span class="p">]</span>
    <span class="c1"># any indel at a specific position in the CDS or PROM (e.g. rpoB@1300_indel</span>
    <span class="c1">#   or rpoB@-15_indel)</span>
    <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

    <span class="c1"># PRIORITY=6: an insertion (or deletion) at a specific position in the CDS or PROM</span>
    <span class="c1">#   (e.g. rpoB@1300_ins or rpoB@-15_del)</span>
    <span class="k">if</span> <span class="n">indel_type</span> <span class="o">!=</span> <span class="s2">&quot;indel&quot;</span> <span class="ow">and</span> <span class="n">indel_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="n">rules_mutation_type_vector</span>
            <span class="o">&amp;</span> <span class="n">rules_position_vector</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">indel_type</span><span class="p">))</span>
        <span class="p">]</span>
        <span class="c1"># any insertion (or deletion) at a specific position in the CDS or PROM</span>
        <span class="c1">#   (e.g. rpoB@1300_ins or rpoB@-15_del)</span>
        <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

    <span class="c1"># PRIORITY=7: an insertion (or deletion) of a specified length at a specific</span>
    <span class="c1">#   position in the CDS or PROM (e.g. rpoB@1300_ins_2 or rpoB@-15_del_200)</span>
    <span class="k">if</span> <span class="n">indel_type</span> <span class="o">!=</span> <span class="s2">&quot;indel&quot;</span> <span class="ow">and</span> <span class="n">indel_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">indel_length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="n">rules_mutation_type_vector</span>
            <span class="o">&amp;</span> <span class="n">rules_position_vector</span>
            <span class="o">&amp;</span> <span class="p">(</span>
                <span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span>
                <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">position</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">indel_type</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">indel_length</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">]</span>
        <span class="c1"># an insertion (or deletion) of a specified length at a specific position in</span>
        <span class="c1">#   the CDS or PROM (e.g. rpoB@1300_ins_2 or rpoB@-15_del_200)</span>
        <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

    <span class="c1"># PRIORITY=8: a frameshifting mutation at a specific position in the CDS</span>
    <span class="c1">#   (e.g. rpoB@100_fs)</span>
    <span class="k">if</span> <span class="n">indel_length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="p">(</span><span class="n">indel_length</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">position</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="n">rules_mutation_type_vector</span>
            <span class="o">&amp;</span> <span class="n">rules_position_vector</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">position</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_fs&quot;</span><span class="p">)</span>
        <span class="p">]</span>
        <span class="c1"># a frameshifting mutation at a specific position in the CDS (e.g. rpoB@100_fs)</span>
        <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

    <span class="c1"># PRIORITY=9: an insertion of a specified series of nucleotides at a position in</span>
    <span class="c1">#   the CDS or PROM (e.g. rpoB@1300_ins_ca)</span>
    <span class="k">if</span> <span class="n">indel_type</span> <span class="o">==</span> <span class="s2">&quot;ins&quot;</span> <span class="ow">and</span> <span class="n">indel_length</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">indel_bases</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
            <span class="n">rules_mutation_type_vector</span>
            <span class="o">&amp;</span> <span class="n">rules_position_vector</span>
            <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">position</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;_ins_&quot;</span> <span class="o">+</span> <span class="n">indel_bases</span><span class="p">)</span>
        <span class="p">]</span>
        <span class="c1"># an insertion of a specified series of nucleotides at a position in the CDS or</span>
        <span class="c1">#   PROM (e.g. rpoB@1300_ins_ca)</span>
        <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.process_snp_variants" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_snp_variants</span><span class="p">(</span><span class="n">mutation_affects</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="n">before</span><span class="p">,</span> <span class="n">after</span><span class="p">,</span> <span class="n">mutation</span><span class="p">,</span> <span class="n">rules</span><span class="p">,</span> <span class="n">rules_mutation_type_vector</span><span class="p">,</span> <span class="n">rules_position_vector</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Apply cascading rules for SNPs according to GARC</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mutation_affects</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Where this mutation affects. i.e PROM/CDS</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>predictions</code>
            </td>
            <td>
                  <code>{int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>str}): Predictions dictionary mapping priority -&gt; prediction.
This is updated for implict return</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>before</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Reference base/AA</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>after</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Alt base/AA</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mutation</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mutation in GARC</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rules</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Rules DataFrame</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rules_mutation_type_vector</code>
            </td>
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Series relating to rules for this
type of mutation</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>rules_position_vector</code>
            </td>
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Series relating to rules for this
position</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>minor</code>
            </td>
            <td>
                  <code>float | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Float for supporting evidence of minor populations
(or None if not a minor population).</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 913</span>
<span class="normal"> 914</span>
<span class="normal"> 915</span>
<span class="normal"> 916</span>
<span class="normal"> 917</span>
<span class="normal"> 918</span>
<span class="normal"> 919</span>
<span class="normal"> 920</span>
<span class="normal"> 921</span>
<span class="normal"> 922</span>
<span class="normal"> 923</span>
<span class="normal"> 924</span>
<span class="normal"> 925</span>
<span class="normal"> 926</span>
<span class="normal"> 927</span>
<span class="normal"> 928</span>
<span class="normal"> 929</span>
<span class="normal"> 930</span>
<span class="normal"> 931</span>
<span class="normal"> 932</span>
<span class="normal"> 933</span>
<span class="normal"> 934</span>
<span class="normal"> 935</span>
<span class="normal"> 936</span>
<span class="normal"> 937</span>
<span class="normal"> 938</span>
<span class="normal"> 939</span>
<span class="normal"> 940</span>
<span class="normal"> 941</span>
<span class="normal"> 942</span>
<span class="normal"> 943</span>
<span class="normal"> 944</span>
<span class="normal"> 945</span>
<span class="normal"> 946</span>
<span class="normal"> 947</span>
<span class="normal"> 948</span>
<span class="normal"> 949</span>
<span class="normal"> 950</span>
<span class="normal"> 951</span>
<span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_snp_variants</span><span class="p">(</span>
    <span class="n">mutation_affects</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">predictions</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">]],</span>
    <span class="n">before</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">after</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">mutation</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">rules</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">rules_mutation_type_vector</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span>
    <span class="n">rules_position_vector</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span>
    <span class="n">minor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Apply cascading rules for SNPs according to GARC</span>

<span class="sd">    Args:</span>
<span class="sd">        mutation_affects (str): Where this mutation affects. i.e PROM/CDS</span>
<span class="sd">        predictions ({int: str}): Predictions dictionary mapping priority -&gt; prediction.</span>
<span class="sd">            This is updated for implict return</span>
<span class="sd">        before (str): Reference base/AA</span>
<span class="sd">        after (str): Alt base/AA</span>
<span class="sd">        mutation (str): Mutation in GARC</span>
<span class="sd">        rules (pandas.DataFrame): Rules DataFrame</span>
<span class="sd">        rules_mutation_type_vector (pandas.Series): Series relating to rules for this</span>
<span class="sd">            type of mutation</span>
<span class="sd">        rules_position_vector (pandas.Series): Series relating to rules for this</span>
<span class="sd">            position</span>
<span class="sd">        minor (float | None): Float for supporting evidence of minor populations</span>
<span class="sd">            (or None if not a minor population).</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">before</span> <span class="o">==</span> <span class="n">after</span><span class="p">:</span>
        <span class="c1"># PRIORITY=0: no change, i.e. wildtype specified</span>

        <span class="c1"># PRIORITY=1: synonymous mutation at any position in the CDS (e.g. rpoB_*=)</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rules_mutation_type_vector</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;*=&quot;</span><span class="p">)]</span>
        <span class="c1"># syn SNP at any position in the CDS</span>
        <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">before</span> <span class="o">!=</span> <span class="n">after</span><span class="p">:</span>
        <span class="c1"># PRIORITY=2: nonsynoymous mutation at any position in the CDS or PROM</span>
        <span class="c1">#   (e.g. rpoB_*? or rpoB_-*?)</span>
        <span class="k">if</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">]:</span>
            <span class="c1"># Don&#39;t let null calls hit default rules</span>
            <span class="k">if</span> <span class="n">mutation_affects</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rules_mutation_type_vector</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;*?&quot;</span><span class="p">)]</span>
                <span class="c1"># nonsyn SNP at any position in the CDS</span>
                <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rules_mutation_type_vector</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;-*?&quot;</span><span class="p">)]</span>
                <span class="c1"># nonsyn SNP at any position in the PROM</span>
                <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
        <span class="c1"># PRIORITY=3: het mutation at any position in the CDS or PROM (e.g. rpoB@*Z</span>
        <span class="c1">#   or rpoB@-*z)</span>
        <span class="k">if</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;Z&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span> <span class="s2">&quot;O&quot;</span><span class="p">,</span> <span class="s2">&quot;o&quot;</span><span class="p">,</span> <span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">mutation_affects</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Z&quot;</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                        <span class="p">(</span><span class="n">rules_mutation_type_vector</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;*Z&quot;</span><span class="p">)</span>
                    <span class="p">]</span>
                    <span class="c1"># het SNP at any position in the CDS</span>
                    <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;O&quot;</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                        <span class="p">(</span><span class="n">rules_mutation_type_vector</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;*O&quot;</span><span class="p">)</span>
                    <span class="p">]</span>
                    <span class="c1"># filter fail at any position in the CDS</span>
                    <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;X&quot;</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                        <span class="p">(</span><span class="n">rules_mutation_type_vector</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;*X&quot;</span><span class="p">)</span>
                    <span class="p">]</span>
                    <span class="c1"># null at any position in the CDS</span>
                    <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;z&quot;</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                        <span class="n">rules_mutation_type_vector</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;-*z&quot;</span><span class="p">)</span>
                    <span class="p">]</span>
                    <span class="c1"># het SNP at any position in the PROM</span>
                    <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;o&quot;</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                        <span class="n">rules_mutation_type_vector</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;-*o&quot;</span><span class="p">)</span>
                    <span class="p">]</span>
                    <span class="c1"># filter fail at any position in the PROM</span>
                    <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;x&quot;</span><span class="p">:</span>
                    <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                        <span class="n">rules_mutation_type_vector</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;-*x&quot;</span><span class="p">)</span>
                    <span class="p">]</span>
                    <span class="c1"># null at any position in the PROM</span>
                    <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

        <span class="c1"># PRIORITY=4: specific mutation at any position in the CDS: only Stop codons</span>
        <span class="c1">#   make sense for now (e.g. rpoB@*!)</span>
        <span class="k">if</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;!&quot;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">mutation_affects</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span><span class="p">:</span>
                <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">rules_mutation_type_vector</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="s2">&quot;*!&quot;</span><span class="p">)]</span>
                <span class="c1"># stop codon at any position in the CDS</span>
                <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

        <span class="c1"># PRIORITY=5: no change at specific position</span>
        <span class="c1"># PRIORITY=6: synonymous mutations at specific location (usually picked up by</span>
        <span class="c1">#   e.g. L202L, rather than L202=)</span>

        <span class="c1"># PRIORITY=7: any het mutation at this specific position in the CDS or PROM</span>
        <span class="c1">#   (e.g. rpoB@S450Z or rpoB@c-15z)</span>
        <span class="k">if</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;Z&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">]:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                <span class="p">(</span><span class="n">rules_mutation_type_vector</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules_position_vector</span><span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;Z&quot;</span><span class="p">,</span> <span class="s2">&quot;z&quot;</span><span class="p">]))</span>
            <span class="p">]</span>
            <span class="c1"># het SNP at specified position in the CDS</span>
            <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

        <span class="c1"># PRIORITY=8: any nonsynoymous mutation at this specific position in the CDS or</span>
        <span class="c1">#   PROM  (e.g. rpoB@S450? or rpoB@c-15?)</span>
        <span class="k">if</span> <span class="n">mutation</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;x&quot;</span><span class="p">]:</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
                <span class="n">rules_mutation_type_vector</span>
                <span class="o">&amp;</span> <span class="n">rules_position_vector</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;?&quot;</span><span class="p">)</span>
            <span class="p">]</span>
            <span class="c1"># nonsyn SNP at specified position in the CDS</span>
            <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>

    <span class="c1"># PRIORITY=9: an exact match</span>
    <span class="n">row</span> <span class="o">=</span> <span class="n">rules</span><span class="o">.</span><span class="n">loc</span><span class="p">[(</span><span class="n">rules_mutation_type_vector</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">rules</span><span class="o">.</span><span class="n">MUTATION</span> <span class="o">==</span> <span class="n">mutation</span><span class="p">)]</span>
    <span class="c1"># exact SNP match</span>
    <span class="n">row_prediction</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.row_prediction" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">row_prediction</span><span class="p">(</span><span class="n">rows</span><span class="p">,</span> <span class="n">predictions</span><span class="p">,</span> <span class="n">priority</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Get the predictions from the catalogue for the applicable rows</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>rows</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>DataFrame of the rows within the catalogue</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>predictions</code>
            </td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Predictions dict mapping priorities to predictions. This is
updated for implict return</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>priority</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Priority of this mutation type</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>minor</code>
            </td>
            <td>
                  <code>float | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Reads/FRS supporting this, or None if not a
minor population</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">row_prediction</span><span class="p">(</span>
    <span class="n">rows</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">predictions</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Dict</span><span class="p">]],</span>
    <span class="n">priority</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">minor</span><span class="p">:</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the predictions from the catalogue for the applicable rows</span>

<span class="sd">    Args:</span>
<span class="sd">        rows (pandas.DataFrame): DataFrame of the rows within the catalogue</span>
<span class="sd">        predictions (dict): Predictions dict mapping priorities to predictions. This is</span>
<span class="sd">            updated for implict return</span>
<span class="sd">        priority (int): Priority of this mutation type</span>
<span class="sd">        minor (float | None): Reads/FRS supporting this, or None if not a</span>
<span class="sd">            minor population</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">pred</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">evidence</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="n">values</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;R&quot;</span><span class="p">,</span> <span class="s2">&quot;U&quot;</span><span class="p">,</span> <span class="s2">&quot;F&quot;</span><span class="p">,</span> <span class="s2">&quot;S&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">rows</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">row</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">int</span><span class="p">(</span><span class="n">priority</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span>
                <span class="mi">1</span><span class="p">,</span> <span class="mi">11</span>
            <span class="p">),</span> <span class="s2">&quot;priority must be an integer in range 1,2..10&quot;</span>
            <span class="k">if</span> <span class="n">values</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">])</span> <span class="o">&lt;</span> <span class="n">values</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">pred</span><span class="p">):</span>
                <span class="c1"># This row&#39;s prediction is more important than the current, so check</span>
                <span class="c1">#   for minors and prioritise</span>
                <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                    <span class="c1"># Neither are minor populations so act normally</span>
                    <span class="n">pred</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">]</span>
                    <span class="n">evidence</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">]</span>

                <span class="k">elif</span> <span class="p">(</span>
                    <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                    <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span>
                    <span class="ow">and</span> <span class="n">minor</span> <span class="o">&lt;</span> <span class="mi">1</span>
                    <span class="ow">and</span> <span class="nb">float</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">1</span>
                <span class="p">):</span>
                    <span class="c1"># We have FRS</span>
                    <span class="k">if</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="nb">float</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]):</span>
                        <span class="c1"># Match</span>
                        <span class="n">pred</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">]</span>
                        <span class="n">evidence</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">]</span>

                <span class="k">elif</span> <span class="p">(</span>
                    <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                    <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span>
                    <span class="ow">and</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="mi">1</span>
                    <span class="ow">and</span> <span class="nb">float</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">])</span> <span class="o">&gt;=</span> <span class="mi">1</span>
                <span class="p">):</span>
                    <span class="c1"># We have COV</span>
                    <span class="k">if</span> <span class="n">minor</span> <span class="o">&gt;=</span> <span class="nb">float</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]):</span>
                        <span class="c1"># Match</span>
                        <span class="n">pred</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;PREDICTION&quot;</span><span class="p">]</span>
                        <span class="n">evidence</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;EVIDENCE&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">pred</span><span class="p">:</span>
        <span class="n">predictions</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">priority</span><span class="p">)]</span> <span class="o">=</span> <span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">pred</span><span class="p">),</span> <span class="n">evidence</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.sanity_check_snp" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">sanity_check_snp</span><span class="p">(</span><span class="n">before</span><span class="p">,</span> <span class="n">after</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Sanity check that a given SNP is valid. i.e check that the bases are valid</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>before</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Reference bases/AA</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>after</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Alt bases/AA</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>AssertationError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Raised in cases in which the SNP is invalid.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1368</span>
<span class="normal">1369</span>
<span class="normal">1370</span>
<span class="normal">1371</span>
<span class="normal">1372</span>
<span class="normal">1373</span>
<span class="normal">1374</span>
<span class="normal">1375</span>
<span class="normal">1376</span>
<span class="normal">1377</span>
<span class="normal">1378</span>
<span class="normal">1379</span>
<span class="normal">1380</span>
<span class="normal">1381</span>
<span class="normal">1382</span>
<span class="normal">1383</span>
<span class="normal">1384</span>
<span class="normal">1385</span>
<span class="normal">1386</span>
<span class="normal">1387</span>
<span class="normal">1388</span>
<span class="normal">1389</span>
<span class="normal">1390</span>
<span class="normal">1391</span>
<span class="normal">1392</span>
<span class="normal">1393</span>
<span class="normal">1394</span>
<span class="normal">1395</span>
<span class="normal">1396</span>
<span class="normal">1397</span>
<span class="normal">1398</span>
<span class="normal">1399</span>
<span class="normal">1400</span>
<span class="normal">1401</span>
<span class="normal">1402</span>
<span class="normal">1403</span>
<span class="normal">1404</span>
<span class="normal">1405</span>
<span class="normal">1406</span>
<span class="normal">1407</span>
<span class="normal">1408</span>
<span class="normal">1409</span>
<span class="normal">1410</span>
<span class="normal">1411</span>
<span class="normal">1412</span>
<span class="normal">1413</span>
<span class="normal">1414</span>
<span class="normal">1415</span>
<span class="normal">1416</span>
<span class="normal">1417</span>
<span class="normal">1418</span>
<span class="normal">1419</span>
<span class="normal">1420</span>
<span class="normal">1421</span>
<span class="normal">1422</span>
<span class="normal">1423</span>
<span class="normal">1424</span>
<span class="normal">1425</span>
<span class="normal">1426</span>
<span class="normal">1427</span>
<span class="normal">1428</span>
<span class="normal">1429</span>
<span class="normal">1430</span>
<span class="normal">1431</span>
<span class="normal">1432</span>
<span class="normal">1433</span>
<span class="normal">1434</span>
<span class="normal">1435</span>
<span class="normal">1436</span>
<span class="normal">1437</span>
<span class="normal">1438</span>
<span class="normal">1439</span>
<span class="normal">1440</span>
<span class="normal">1441</span>
<span class="normal">1442</span>
<span class="normal">1443</span>
<span class="normal">1444</span>
<span class="normal">1445</span>
<span class="normal">1446</span>
<span class="normal">1447</span>
<span class="normal">1448</span>
<span class="normal">1449</span>
<span class="normal">1450</span>
<span class="normal">1451</span>
<span class="normal">1452</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">sanity_check_snp</span><span class="p">(</span><span class="n">before</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span> <span class="n">after</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Sanity check that a given SNP is valid. i.e check that the bases are valid</span>

<span class="sd">    Args:</span>
<span class="sd">        before (str): Reference bases/AA</span>
<span class="sd">        after (str): Alt bases/AA</span>

<span class="sd">    Raises:</span>
<span class="sd">        AssertationError: Raised in cases in which the SNP is invalid.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">assert</span> <span class="n">after</span> <span class="ow">in</span> <span class="p">[</span>
        <span class="s2">&quot;a&quot;</span><span class="p">,</span>
        <span class="s2">&quot;c&quot;</span><span class="p">,</span>
        <span class="s2">&quot;t&quot;</span><span class="p">,</span>
        <span class="s2">&quot;g&quot;</span><span class="p">,</span>
        <span class="s2">&quot;x&quot;</span><span class="p">,</span>
        <span class="s2">&quot;z&quot;</span><span class="p">,</span>
        <span class="s2">&quot;o&quot;</span><span class="p">,</span>
        <span class="s2">&quot;?&quot;</span><span class="p">,</span>
        <span class="s2">&quot;!&quot;</span><span class="p">,</span>
        <span class="s2">&quot;A&quot;</span><span class="p">,</span>
        <span class="s2">&quot;C&quot;</span><span class="p">,</span>
        <span class="s2">&quot;D&quot;</span><span class="p">,</span>
        <span class="s2">&quot;E&quot;</span><span class="p">,</span>
        <span class="s2">&quot;F&quot;</span><span class="p">,</span>
        <span class="s2">&quot;G&quot;</span><span class="p">,</span>
        <span class="s2">&quot;H&quot;</span><span class="p">,</span>
        <span class="s2">&quot;I&quot;</span><span class="p">,</span>
        <span class="s2">&quot;K&quot;</span><span class="p">,</span>
        <span class="s2">&quot;L&quot;</span><span class="p">,</span>
        <span class="s2">&quot;M&quot;</span><span class="p">,</span>
        <span class="s2">&quot;N&quot;</span><span class="p">,</span>
        <span class="s2">&quot;O&quot;</span><span class="p">,</span>
        <span class="s2">&quot;P&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Q&quot;</span><span class="p">,</span>
        <span class="s2">&quot;R&quot;</span><span class="p">,</span>
        <span class="s2">&quot;S&quot;</span><span class="p">,</span>
        <span class="s2">&quot;T&quot;</span><span class="p">,</span>
        <span class="s2">&quot;V&quot;</span><span class="p">,</span>
        <span class="s2">&quot;W&quot;</span><span class="p">,</span>
        <span class="s2">&quot;X&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Y&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Z&quot;</span><span class="p">,</span>
    <span class="p">],</span> <span class="p">(</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">after</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; is not a recognised amino acid or base!&quot;</span>
    <span class="p">)</span>
    <span class="k">assert</span> <span class="n">before</span> <span class="ow">in</span> <span class="p">[</span>
        <span class="s2">&quot;a&quot;</span><span class="p">,</span>
        <span class="s2">&quot;c&quot;</span><span class="p">,</span>
        <span class="s2">&quot;t&quot;</span><span class="p">,</span>
        <span class="s2">&quot;g&quot;</span><span class="p">,</span>
        <span class="s2">&quot;?&quot;</span><span class="p">,</span>
        <span class="s2">&quot;!&quot;</span><span class="p">,</span>
        <span class="s2">&quot;A&quot;</span><span class="p">,</span>
        <span class="s2">&quot;C&quot;</span><span class="p">,</span>
        <span class="s2">&quot;D&quot;</span><span class="p">,</span>
        <span class="s2">&quot;E&quot;</span><span class="p">,</span>
        <span class="s2">&quot;F&quot;</span><span class="p">,</span>
        <span class="s2">&quot;G&quot;</span><span class="p">,</span>
        <span class="s2">&quot;H&quot;</span><span class="p">,</span>
        <span class="s2">&quot;I&quot;</span><span class="p">,</span>
        <span class="s2">&quot;K&quot;</span><span class="p">,</span>
        <span class="s2">&quot;L&quot;</span><span class="p">,</span>
        <span class="s2">&quot;M&quot;</span><span class="p">,</span>
        <span class="s2">&quot;N&quot;</span><span class="p">,</span>
        <span class="s2">&quot;P&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Q&quot;</span><span class="p">,</span>
        <span class="s2">&quot;R&quot;</span><span class="p">,</span>
        <span class="s2">&quot;S&quot;</span><span class="p">,</span>
        <span class="s2">&quot;T&quot;</span><span class="p">,</span>
        <span class="s2">&quot;V&quot;</span><span class="p">,</span>
        <span class="s2">&quot;W&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Y&quot;</span><span class="p">,</span>
    <span class="p">],</span> <span class="p">(</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">before</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot; is not a recognised amino acid or base!&quot;</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">before</span><span class="o">.</span><span class="n">islower</span><span class="p">():</span>
        <span class="k">assert</span> <span class="n">after</span><span class="o">.</span><span class="n">islower</span><span class="p">(),</span> <span class="s2">&quot;nucleotides must be lowercase!&quot;</span>
        <span class="k">assert</span> <span class="p">(</span>
            <span class="n">before</span> <span class="o">!=</span> <span class="n">after</span>
        <span class="p">),</span> <span class="s2">&quot;makes no sense for the nucleotide to be the same in a mutation!&quot;</span>
    <span class="k">elif</span> <span class="n">before</span><span class="o">.</span><span class="n">isupper</span><span class="p">():</span>
        <span class="k">assert</span> <span class="n">after</span><span class="o">.</span><span class="n">isupper</span><span class="p">()</span> <span class="ow">or</span> <span class="n">after</span> <span class="o">==</span> <span class="s2">&quot;!&quot;</span><span class="p">,</span> <span class="s2">&quot;amino acids must be UPPERCASE!&quot;</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.split_mutation" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">split_mutation</span><span class="p">(</span><span class="n">row</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Take a row of the catalogue and get info about the mutation from it.
This includes the type, ref, alt, position and minor populations</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>row</code>
            </td>
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Catalogue row</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Raised when mutations are malformed</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.Series">Series</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pandas.Series: Series about the mutation. In order: gene, mutation, position,
type of position (i.e PROM/CDS), mutation type (i.e INDEL/SNP/MULTI),
support for minor populations</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">split_mutation</span><span class="p">(</span><span class="n">row</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Take a row of the catalogue and get info about the mutation from it.</span>
<span class="sd">    This includes the type, ref, alt, position and minor populations</span>

<span class="sd">    Args:</span>
<span class="sd">        row (pandas.Series): Catalogue row</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: Raised when mutations are malformed</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.Series: Series about the mutation. In order: gene, mutation, position,</span>
<span class="sd">            type of position (i.e PROM/CDS), mutation type (i.e INDEL/SNP/MULTI),</span>
<span class="sd">            support for minor populations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Strip any whitespace as this can cause issues</span>
    <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
    <span class="k">if</span> <span class="s2">&quot;&amp;&quot;</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]:</span>
        <span class="c1"># Multi-mutation so don&#39;t try to split it out in the same way</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="s2">&quot;MULTI&quot;</span>

        <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;^&quot;</span><span class="p">:</span>
            <span class="n">mutation_type</span> <span class="o">=</span> <span class="s2">&quot;EPISTASIS&quot;</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">::]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mutation_type</span> <span class="o">=</span> <span class="s2">&quot;MULTI&quot;</span>

        <span class="c1"># Ensure mutations are in a reproducable order</span>
        <span class="n">mutation</span> <span class="o">=</span> <span class="s2">&quot;&amp;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">))))</span>

        <span class="n">position</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">mutation_affects</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="c1"># Check for minor populations (and remove, we&#39;ll check these separately)</span>
        <span class="n">minors</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="p">[</span><span class="n">mut</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">mut</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span> <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">)]</span>
        <span class="p">)</span>

        <span class="n">mutation</span> <span class="o">=</span> <span class="s2">&quot;&amp;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">mut</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">)])</span>

        <span class="c1"># Check for validity</span>
        <span class="n">validate_multi</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
            <span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="n">mutation</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">mutation_affects</span><span class="p">,</span> <span class="n">mutation_type</span><span class="p">,</span> <span class="n">minors</span><span class="p">]</span>
        <span class="p">)</span>

    <span class="c1"># split the supplied mutation on _ which separates the components</span>
    <span class="n">components</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;@&quot;</span><span class="p">)</span>

    <span class="c1"># the gene name is always the first component</span>
    <span class="n">gene</span> <span class="o">=</span> <span class="n">components</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># ..and the remainder is the mutation</span>
    <span class="n">mutation</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">gene</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>

    <span class="c1"># Check for minor populations (and remove minors, we&#39;ll check for those separately)</span>
    <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">:</span>
        <span class="n">minor</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">mutation</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">minor</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="n">cols</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="n">valid</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">mutation_type</span> <span class="o">=</span> <span class="s2">&quot;SNP&quot;</span>
        <span class="n">mutation_affects</span> <span class="o">=</span> <span class="s2">&quot;CDS&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">position</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">position</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">mutation_affects</span> <span class="o">=</span> <span class="s2">&quot;PROM&quot;</span>
            <span class="k">elif</span> <span class="nb">int</span><span class="p">(</span><span class="n">position</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">valid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="n">position</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">][:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">position</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
                <span class="n">mutation_affects</span> <span class="o">=</span> <span class="s2">&quot;PROM&quot;</span>

    <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">]:</span>
        <span class="n">mutation_type</span> <span class="o">=</span> <span class="s2">&quot;INDEL&quot;</span>
        <span class="n">percentage</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
            <span class="c1"># Checking for percentage deletions (of form &quot;del_&lt;percent&gt;&quot;)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="nb">float</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">percentage</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="n">percentage</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="n">percentage</span><span class="p">:</span>
                <span class="n">position</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">percentage</span><span class="p">:</span>
            <span class="n">mutation_affects</span> <span class="o">=</span> <span class="s2">&quot;GENE&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mutation_affects</span> <span class="o">=</span> <span class="s2">&quot;CDS&quot;</span>
            <span class="n">position</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">position</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">mutation_affects</span> <span class="o">=</span> <span class="s2">&quot;PROM&quot;</span>
                <span class="k">elif</span> <span class="nb">int</span><span class="p">(</span><span class="n">position</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">valid</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">position</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>
                        <span class="n">mutation_affects</span> <span class="o">=</span> <span class="s2">&quot;PROM&quot;</span>
                <span class="k">except</span> <span class="ne">IndexError</span><span class="p">:</span>
                    <span class="n">valid</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Badly formed mutation: &quot;</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">])</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">valid</span><span class="p">:</span>
        <span class="c1"># Valid mutation not found, so complain</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Badly formed mutation: &quot;</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
        <span class="p">[</span><span class="n">gene</span><span class="p">,</span> <span class="n">mutation</span><span class="p">,</span> <span class="n">position</span><span class="p">,</span> <span class="n">mutation_affects</span><span class="p">,</span> <span class="n">mutation_type</span><span class="p">,</span> <span class="n">minor</span><span class="p">]</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="grammar_GARC1.validate_multi" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate_multi</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Validate that we have a valid multi-mutation.</p>
<p>This checks that if a multi contains specific mutations, they should not be covered
by general mutations which this multi includes.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mutation</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Multi-mutation</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>piezo/grammar_GARC1.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate_multi</span><span class="p">(</span><span class="n">mutation</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Validate that we have a valid multi-mutation.</span>

<span class="sd">    This checks that if a multi contains specific mutations, they should not be covered</span>
<span class="sd">    by general mutations which this multi includes.</span>

<span class="sd">    Args:</span>
<span class="sd">        mutation (str): Multi-mutation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># If we have a specific rule, combined with a general rule which covers it,</span>
    <span class="c1">#   that should be reduced, so it&#39;s invalid.</span>
    <span class="c1"># If we were to allow defaults which cover specifics, the multi matching would break</span>
    <span class="c1">#</span>
    <span class="c1"># Easiest way to do this is to take general rules and construct a catalogue</span>
    <span class="c1">#   from them, then, pass in all specific rules, and if we get a match on any of</span>
    <span class="c1">#   them, they are covered by a general rule</span>
    <span class="n">rule_mutations</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">)</span>
    <span class="n">dummy_cat</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;EVIDENCE&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="n">ujson</span><span class="o">.</span><span class="n">dumps</span><span class="p">({</span><span class="s2">&quot;Rule hit&quot;</span><span class="p">:</span> <span class="n">idx</span><span class="p">})</span> <span class="k">for</span> <span class="p">(</span><span class="n">idx</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">rule_mutations</span><span class="p">)</span>
        <span class="p">],</span>
        <span class="s2">&quot;MUTATION&quot;</span><span class="p">:</span> <span class="n">rule_mutations</span><span class="p">,</span>
        <span class="s2">&quot;DRUG&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;na&quot;</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">rule_mutations</span><span class="p">],</span>
        <span class="s2">&quot;PREDICTION&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;R&quot;</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">rule_mutations</span><span class="p">],</span>
    <span class="p">}</span>
    <span class="n">r</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">dummy_cat</span><span class="p">)</span>
    <span class="n">r</span><span class="p">[</span>
        <span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">,</span> <span class="s2">&quot;MUTATION&quot;</span><span class="p">,</span> <span class="s2">&quot;POSITION&quot;</span><span class="p">,</span> <span class="s2">&quot;MUTATION_AFFECTS&quot;</span><span class="p">,</span> <span class="s2">&quot;MUTATION_TYPE&quot;</span><span class="p">,</span> <span class="s2">&quot;MINOR&quot;</span><span class="p">]</span>
    <span class="p">]</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">split_mutation</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">generals</span> <span class="o">=</span> <span class="n">r</span><span class="p">[(</span><span class="n">r</span><span class="p">[</span><span class="s2">&quot;POSITION&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;*&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\?&quot;</span><span class="p">))]</span>
    <span class="n">specifics</span> <span class="o">=</span> <span class="n">r</span><span class="p">[(</span><span class="n">r</span><span class="p">[</span><span class="s2">&quot;POSITION&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;*&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="o">~</span><span class="n">r</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;\?&quot;</span><span class="p">))]</span>
    <span class="c1"># As this is a named **tuple**, it is immutable so create a new one</span>
    <span class="n">cat</span> <span class="o">=</span> <span class="n">Catalogue</span><span class="p">(</span>
        <span class="s2">&quot;NC_000962.3&quot;</span><span class="p">,</span>
        <span class="s2">&quot;generic&quot;</span><span class="p">,</span>
        <span class="s2">&quot;1.0&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GARC1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;R&quot;</span><span class="p">,</span>  <span class="c1"># Obviously not R but RFUS are needed here</span>
        <span class="p">[</span><span class="s2">&quot;na&quot;</span><span class="p">],</span>
        <span class="n">generals</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
        <span class="p">{</span><span class="s2">&quot;na&quot;</span><span class="p">:</span> <span class="n">generals</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()},</span>
        <span class="p">{</span><span class="n">gene</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;na&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">generals</span><span class="p">[</span><span class="s2">&quot;GENE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()},</span>
        <span class="nb">len</span><span class="p">(</span><span class="n">generals</span><span class="p">),</span>
        <span class="n">generals</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">specifics</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">pred</span> <span class="o">=</span> <span class="n">predict_GARC1</span><span class="p">(</span><span class="n">cat</span><span class="p">,</span> <span class="s2">&quot;@&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">[[</span><span class="s2">&quot;GENE&quot;</span><span class="p">,</span> <span class="s2">&quot;MUTATION&quot;</span><span class="p">]]))</span>
            <span class="k">if</span> <span class="n">pred</span> <span class="o">!=</span> <span class="s2">&quot;S&quot;</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Badly formed mutation: &quot;</span>
                    <span class="o">+</span> <span class="n">mutation</span>
                    <span class="o">+</span> <span class="s2">&quot; contains generic rules which cover specific rules!&quot;</span>
                <span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">if</span> <span class="s2">&quot;Badly formed mutation: &quot;</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">):</span>
                <span class="c1"># If we&#39;re catching the exception we just threw, re-throw it</span>
                <span class="k">raise</span> <span class="n">e</span>
            <span class="k">continue</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../catalogue/" class="md-footer__link md-footer__link--prev" aria-label="Previous: catalogue">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                catalogue
              </div>
            </div>
          </a>
        
        
          
          <a href="../../nomenclature/" class="md-footer__link md-footer__link--next" aria-label="Next: Nomenclature">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                Nomenclature
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.footer", "navigation.instant", "navigation.sections", "navigation.tabs", "navigation.tracking"], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "jeremy"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.f1b6f286.min.js"></script>
      
    
  </body>
</html>